

<!DOCTYPE html>
<html lang="zh-CN" data-default-color-scheme=auto>



<head>
  <meta charset="UTF-8">

  <link rel="apple-touch-icon" sizes="76x76" href="/img/favicon-32x32.png">
  <link rel="icon" href="/img/favicon-32x32.png">
  

  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=5.0, shrink-to-fit=no">
  <meta http-equiv="x-ua-compatible" content="ie=edge">
  
  <meta name="theme-color" content="#2f4154">
  <meta name="author" content="Jingxi Qin">
  <meta name="keywords" content="">
  
    <meta name="description" content="准备Seurat对象Read10X和ReadRDS成功过，txt和csv还没试过 123456789library(dplyr)library(Seurat)library(patchwork)# Load the PBMC datasetpbmc.data &lt;- Read10X(data.dir &#x3D; &quot;&#x2F;brahms&#x2F;mollag&#x2F;practice&#x2F;filtered_gene_b">
<meta property="og:type" content="article">
<meta property="og:title" content="Learning Seurat - Guided Clustering Tutorial">
<meta property="og:url" content="http://example.com/2024/09/03/20240903Seurat%20-%20Guided%20Clustering%20Tutorial/index.html">
<meta property="og:site_name" content="Jingxi Qin Blog">
<meta property="og:description" content="准备Seurat对象Read10X和ReadRDS成功过，txt和csv还没试过 123456789library(dplyr)library(Seurat)library(patchwork)# Load the PBMC datasetpbmc.data &lt;- Read10X(data.dir &#x3D; &quot;&#x2F;brahms&#x2F;mollag&#x2F;practice&#x2F;filtered_gene_b">
<meta property="og:locale" content="zh_CN">
<meta property="og:image" content="http://example.com/image/Seuratpictures%5Cqc2-1.png">
<meta property="og:image" content="http://example.com/image/Seuratpictures%5Cqc2-2.png">
<meta property="og:image" content="http://example.com/image/Seuratpictures%5C1.png">
<meta property="og:image" content="http://example.com/image/Seuratpictures%5Cvar_features-1.png">
<meta property="og:image" content="http://example.com/image/Seuratpictures%5CPCA1.png">
<meta property="og:image" content="http://example.com/image/Seuratpictures%5CPCA2.png">
<meta property="og:image" content="http://example.com/image/Seuratpictures%5Cpca_viz-1.png">
<meta property="og:image" content="http://example.com/image/Seuratpictures%5Cpca_viz-2.png">
<meta property="og:image" content="http://example.com/image/Seuratpictures%5Csingle-heatmap-1.png">
<meta property="og:image" content="http://example.com/image/Seuratpictures%5Celbow_plot-1.png">
<meta property="og:image" content="http://example.com/image/Seuratpictures%5Cumapplot-1.png">
<meta property="og:image" content="http://example.com/image/Seuratpictures%5Cumap-resolution=2.png">
<meta property="og:image" content="http://example.com/image/Seuratpictures%5Cumap-resolution=0.8.png">
<meta property="og:image" content="http://example.com/image/Seuratpictures%5Clog2FoldChange.png">
<meta property="og:image" content="http://example.com/image/Seuratpictures%5Cunnamed-chunk-3-1.png">
<meta property="og:image" content="http://example.com/image/Seuratpictures%5CclusterHeatmap-1.png">
<meta property="og:image" content="http://example.com/image/Seuratpictures%5Clabelplot-1.png">
<meta property="article:published_time" content="2024-09-03T05:15:00.000Z">
<meta property="article:modified_time" content="2024-09-03T05:22:56.829Z">
<meta property="article:author" content="Jingxi Qin">
<meta property="article:tag" content="R">
<meta property="article:tag" content="Seurat">
<meta name="twitter:card" content="summary_large_image">
<meta name="twitter:image" content="http://example.com/image/Seuratpictures%5Cqc2-1.png">
  
  
    <meta name="referrer" content="no-referrer-when-downgrade">
  
  
  <title>Learning Seurat - Guided Clustering Tutorial - Jingxi Qin Blog</title>

  <link  rel="stylesheet" href="https://lib.baomitu.com/twitter-bootstrap/4.6.1/css/bootstrap.min.css" />



  <link  rel="stylesheet" href="https://lib.baomitu.com/github-markdown-css/4.0.0/github-markdown.min.css" />

  <link  rel="stylesheet" href="https://lib.baomitu.com/hint.css/2.7.0/hint.min.css" />

  <link  rel="stylesheet" href="https://lib.baomitu.com/fancybox/3.5.7/jquery.fancybox.min.css" />



<!-- 主题依赖的图标库，不要自行修改 -->
<!-- Do not modify the link that theme dependent icons -->

<link rel="stylesheet" href="//at.alicdn.com/t/c/font_1749284_5i9bdhy70f8.css">



<link rel="stylesheet" href="//at.alicdn.com/t/c/font_1736178_k526ubmyhba.css">


<link  rel="stylesheet" href="/css/main.css" />


  <link id="highlight-css" rel="stylesheet" href="/css/highlight.css" />
  
    <link id="highlight-css-dark" rel="stylesheet" href="/css/highlight-dark.css" />
  




  <script id="fluid-configs">
    var Fluid = window.Fluid || {};
    Fluid.ctx = Object.assign({}, Fluid.ctx)
    var CONFIG = {"hostname":"example.com","root":"/","version":"1.9.8","typing":{"enable":true,"typeSpeed":70,"cursorChar":"_","loop":false,"scope":[]},"anchorjs":{"enable":true,"element":"h1,h2,h3,h4,h5,h6","placement":"left","visible":"hover","icon":""},"progressbar":{"enable":true,"height_px":3,"color":"#29d","options":{"showSpinner":false,"trickleSpeed":100}},"code_language":{"enable":true,"default":"TEXT"},"copy_btn":true,"image_caption":{"enable":true},"image_zoom":{"enable":true,"img_url_replace":["",""]},"toc":{"enable":true,"placement":"right","headingSelector":"h1,h2,h3,h4,h5,h6","collapseDepth":0},"lazyload":{"enable":true,"loading_img":"/img/loading.gif","onlypost":false,"offset_factor":2},"web_analytics":{"enable":true,"follow_dnt":false,"baidu":null,"google":{"measurement_id":null},"tencent":{"sid":null,"cid":null},"leancloud":{"app_id":null,"app_key":null,"server_url":null,"path":"window.location.pathname","ignore_local":true},"umami":{"src":null,"website_id":null,"domains":null,"start_time":"2024-01-01T00:00:00.000Z","token":null,"api_server":null}},"search_path":"/local-search.xml","include_content_in_search":true};

    if (CONFIG.web_analytics.follow_dnt) {
      var dntVal = navigator.doNotTrack || window.doNotTrack || navigator.msDoNotTrack;
      Fluid.ctx.dnt = dntVal && (dntVal.startsWith('1') || dntVal.startsWith('yes') || dntVal.startsWith('on'));
    }
  </script>
  <script  src="/js/utils.js" ></script>
  <script  src="/js/color-schema.js" ></script>
  

  

  

  

  

  



  
</head>


<body>
  

  <header>
    

<div class="header-inner" style="height: 70vh;">
  <nav id="navbar" class="navbar fixed-top  navbar-expand-lg navbar-dark scrolling-navbar">
  <div class="container">
    <a class="navbar-brand" href="/">
      <strong>Jingxi Qin Blog @ SYSU</strong>
    </a>

    <button id="navbar-toggler-btn" class="navbar-toggler" type="button" data-toggle="collapse"
            data-target="#navbarSupportedContent"
            aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation">
      <div class="animated-icon"><span></span><span></span><span></span></div>
    </button>

    <!-- Collapsible content -->
    <div class="collapse navbar-collapse" id="navbarSupportedContent">
      <ul class="navbar-nav ml-auto text-center">
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/" target="_self">
                <i class="iconfont icon-home-fill"></i>
                <span>首页</span>
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/archives/" target="_self">
                <i class="iconfont icon-archive-fill"></i>
                <span>归档</span>
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/tags/" target="_self">
                <i class="iconfont icon-tags-fill"></i>
                <span>标签</span>
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/about/" target="_self">
                <i class="iconfont icon-user-fill"></i>
                <span>关于</span>
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/links/" target="_self">
                <i class="iconfont icon-link-fill"></i>
                <span>友链</span>
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/new/" target="_self">
                <i class="iconfont icon-books"></i>
                <span>cheatsheet</span>
              </a>
            </li>
          
        
        
          <li class="nav-item" id="search-btn">
            <a class="nav-link" target="_self" href="javascript:;" data-toggle="modal" data-target="#modalSearch" aria-label="Search">
              <i class="iconfont icon-search"></i>
            </a>
          </li>
          
        
        
          <li class="nav-item" id="color-toggle-btn">
            <a class="nav-link" target="_self" href="javascript:;" aria-label="Color Toggle">
              <i class="iconfont icon-dark" id="color-toggle-icon"></i>
            </a>
          </li>
        
      </ul>
    </div>
  </div>
</nav>

  

<div id="banner" class="banner" parallax=true
     style="background: url('/img/default2.png') no-repeat center center; background-size: cover;">
  <div class="full-bg-img">
    <div class="mask flex-center" style="background-color: rgba(0, 0, 0, 0.3)">
      <div class="banner-text text-center fade-in-up">
        <div class="h2">
          
            <span id="subtitle" data-typed-text="Learning Seurat - Guided Clustering Tutorial"></span>
          
        </div>

        
          
  <div class="mt-3">
    
    
      <span class="post-meta">
        <i class="iconfont icon-date-fill" aria-hidden="true"></i>
        <time datetime="2024-09-03 13:15" pubdate>
          2024年9月3日 下午
        </time>
      </span>
    
  </div>

  <div class="mt-1">
    
      <span class="post-meta mr-2">
        <i class="iconfont icon-chart"></i>
        
          6.7k 字
        
      </span>
    

    
      <span class="post-meta mr-2">
        <i class="iconfont icon-clock-fill"></i>
        
        
        
          57 分钟
        
      </span>
    

    
    
  </div>


        
      </div>

      
    </div>
  </div>
</div>

</div>

  </header>

  <main>
    
      

<div class="container-fluid nopadding-x">
  <div class="row nomargin-x">
    <div class="side-col d-none d-lg-block col-lg-2">
      

    </div>

    <div class="col-lg-8 nopadding-x-md">
      <div class="container nopadding-x-md" id="board-ctn">
        <div id="board">
          <article class="post-content mx-auto">
            <h1 id="seo-header">Learning Seurat - Guided Clustering Tutorial</h1>
            
            
              <div class="markdown-body">
                
                <h2 id="准备Seurat对象"><a href="#准备Seurat对象" class="headerlink" title="准备Seurat对象"></a>准备Seurat对象</h2><p>Read10X和ReadRDS成功过，txt和csv还没试过</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><code class="hljs R">library<span class="hljs-punctuation">(</span>dplyr<span class="hljs-punctuation">)</span><br>library<span class="hljs-punctuation">(</span>Seurat<span class="hljs-punctuation">)</span><br>library<span class="hljs-punctuation">(</span>patchwork<span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># Load the PBMC dataset</span><br>pbmc.data <span class="hljs-operator">&lt;-</span> Read10X<span class="hljs-punctuation">(</span>data.dir <span class="hljs-operator">=</span> <span class="hljs-string">&quot;/brahms/mollag/practice/filtered_gene_bc_matrices/hg19/&quot;</span><span class="hljs-punctuation">)</span><br><span class="hljs-comment"># Initialize the Seurat object with the raw (non-normalized data).</span><br>pbmc <span class="hljs-operator">&lt;-</span> CreateSeuratObject<span class="hljs-punctuation">(</span>counts <span class="hljs-operator">=</span> pbmc.data<span class="hljs-punctuation">,</span> project <span class="hljs-operator">=</span> <span class="hljs-string">&quot;pbmc3k&quot;</span><span class="hljs-punctuation">,</span> min.cells <span class="hljs-operator">=</span> <span class="hljs-number">3</span><span class="hljs-punctuation">,</span> min.features <span class="hljs-operator">=</span> <span class="hljs-number">200</span><span class="hljs-punctuation">)</span><br>pbmc<br></code></pre></td></tr></table></figure>

<figure class="highlight clean"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><code class="hljs clean"># Lets examine a few genes <span class="hljs-keyword">in</span> the first thirty cells<br>pbmc.data[c(<span class="hljs-string">&quot;CD3D&quot;</span>, <span class="hljs-string">&quot;TCL1A&quot;</span>, <span class="hljs-string">&quot;MS4A1&quot;</span>), <span class="hljs-number">1</span>:<span class="hljs-number">30</span>]<br>## <span class="hljs-number">3</span> x <span class="hljs-number">30</span> sparse Matrix <span class="hljs-keyword">of</span> <span class="hljs-keyword">class</span> <span class="hljs-string">&quot;dgCMatrix&quot;</span><br>##                                                                    <br>## CD3D  <span class="hljs-number">4</span> . <span class="hljs-number">10</span> . . <span class="hljs-number">1</span> <span class="hljs-number">2</span> <span class="hljs-number">3</span> <span class="hljs-number">1</span> . . <span class="hljs-number">2</span> <span class="hljs-number">7</span> <span class="hljs-number">1</span> . . <span class="hljs-number">1</span> <span class="hljs-number">3</span> . <span class="hljs-number">2</span>  <span class="hljs-number">3</span> . . . . . <span class="hljs-number">3</span> <span class="hljs-number">4</span> <span class="hljs-number">1</span> <span class="hljs-number">5</span><br>## TCL1A . .  . . . . . . <span class="hljs-number">1</span> . . . . . . . . . . .  . <span class="hljs-number">1</span> . . . . . . . .<br>## MS4A1 . <span class="hljs-number">6</span>  . . . . . . <span class="hljs-number">1</span> <span class="hljs-number">1</span> <span class="hljs-number">1</span> . . . . . . . . . <span class="hljs-number">36</span> <span class="hljs-number">1</span> <span class="hljs-number">2</span> . . <span class="hljs-number">2</span> . . . .<br></code></pre></td></tr></table></figure>

<p>以上结果说明：<br>seurat object 数据组织格式——<strong>列为基因名</strong>，此处不是ensembl id，但也有的是用的ensemble id，<strong>行是细胞</strong>，<strong>选的时候先是选行的基因，再是列，选细胞</strong></p>
<p>大多数scRNA-seq矩阵中的数据都是0，这里Seurat是用稀疏矩阵表示的。在R中，sparse-matrix 和matrix是不一样的， 用as.matrix转换，不转换默认为sparse-matrix。下面是稀疏矩阵转为R默认矩阵</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><code class="hljs R">dense.size <span class="hljs-operator">&lt;-</span> object.size<span class="hljs-punctuation">(</span>as.matrix<span class="hljs-punctuation">(</span>pbmc.data<span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span><br>dense.size<br></code></pre></td></tr></table></figure>

<h2 id="一、标准预处理工作流"><a href="#一、标准预处理工作流" class="headerlink" title="一、标准预处理工作流"></a>一、标准预处理工作流</h2><p>预处理工作流包括：</p>
<ul>
<li>根据QC指标（quality control）选择和过滤细胞</li>
<li>规范化数据（normalizing data）</li>
<li>缩放</li>
</ul>
<h3 id="1-质量控制QC和为进一步分析选择细胞"><a href="#1-质量控制QC和为进一步分析选择细胞" class="headerlink" title="1.质量控制QC和为进一步分析选择细胞"></a>1.质量控制QC和为进一步分析选择细胞</h3><p>scRNA-seq关键一步是确保在下游分析中只包括单个活细胞。在单个基因中量化基因表达促进了对全基因组中转录波动（噪音）的研究。</p>
<p>微流控或液滴技术，一次运行中对数以万计的细胞进行测序。现有的scRNA-seq方案导致细胞受到压力、破坏、死亡，或者有空的捕获——低质量的细胞。</p>
<p>与细胞质相关的基因在破碎的细胞中强烈下调，（双侧配对t检验），与高质量细胞相比，破碎的细胞在转录组范围内的噪声水平增加，<strong>含有多个细胞的孔显示出与破碎细胞相似的表达和噪声模式</strong>，这表明多个细胞包含破碎细胞和高质量细胞的混合物。</p>
<ul>
<li>The number of unique genes detected in each cell.<ul>
<li><strong>Low-quality cells</strong> or <strong>empty droplets</strong> will often have very few genes</li>
<li><strong>Cell doublets</strong> or <strong>multiplets</strong> may exhibit an aberrantly high gene count（高了不好，而且不仅仅是高，多个细胞在一个孔里往往存在细胞破坏）</li>
</ul>
</li>
<li>Similarly, the total number of molecules detected within a cell (correlates strongly with unique genes)</li>
<li>The percentage of reads that map to the mitochondrial genome<ul>
<li>Low-quality &#x2F; dying cells often exhibit extensive mitochondrial contamination（有些细胞在镜下和正常可能没有明显不同，但是也有mRNA的泄露，线粒体RNA表达上调。破碎细胞中 mtDNA 的 RNA 水平上调表明细胞质含量丢失。在细胞膜破裂的情况下，细胞质RNA会丢失，但线粒体中封闭的RNA会保留）</li>
<li>We calculate <strong>mitochondrial QC metrics</strong> with the function, which calculates the percentage of counts originating from a set of features<code>PercentageFeatureSet()</code></li>
<li>We use the set of all genes starting with <code>MT-</code>as a set of mitochondrial genes</li>
</ul>
</li>
</ul>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><code class="hljs R"><span class="hljs-comment"># The [[ operator can add columns to object metadata. This is a great place to stash QC stats</span><br>pbmc<span class="hljs-punctuation">[[</span><span class="hljs-string">&quot;percent.mt&quot;</span><span class="hljs-punctuation">]</span><span class="hljs-punctuation">]</span> <span class="hljs-operator">&lt;-</span> PercentageFeatureSet<span class="hljs-punctuation">(</span>pbmc<span class="hljs-punctuation">,</span> pattern <span class="hljs-operator">=</span> <span class="hljs-string">&quot;^MT-&quot;</span><span class="hljs-punctuation">)</span><br></code></pre></td></tr></table></figure>

<p><code>计算给定特征的计数所占的比例：PercentageFeatureSet</code>为每个细胞计算符合某种特征的总数的百分比，比如计算细胞中映射到线粒体转录本中的基因比例。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><code class="hljs R">PercentageFeatureSet<span class="hljs-punctuation">(</span><br>  object<span class="hljs-punctuation">,</span><br>  <span class="hljs-comment"># 正则用于匹配</span><br>  pattern <span class="hljs-operator">=</span> <span class="hljs-literal">NULL</span><span class="hljs-punctuation">,</span><br>  features <span class="hljs-operator">=</span> <span class="hljs-literal">NULL</span><span class="hljs-punctuation">,</span><br>  col.name <span class="hljs-operator">=</span> <span class="hljs-literal">NULL</span><span class="hljs-punctuation">,</span><br>  assay <span class="hljs-operator">=</span> <span class="hljs-literal">NULL</span><br><span class="hljs-punctuation">)</span><br></code></pre></td></tr></table></figure>

<p>QC 标准在Seurat中在<code>CreateSeuratObject()</code>时就已经计算过并保存了，通过meta.data查看</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><code class="hljs R"><span class="hljs-comment"># Show QC metrics for the first 5 cells</span><br>head<span class="hljs-punctuation">(</span>pbmc<span class="hljs-operator">@</span>meta.data<span class="hljs-punctuation">,</span> <span class="hljs-number">5</span><span class="hljs-punctuation">)</span><br><span class="hljs-comment">##                  orig.ident nCount_RNA nFeature_RNA percent.mt</span><br><span class="hljs-comment">## AAACATACAACCAC-1     pbmc3k       2419          779  3.0177759</span><br><span class="hljs-comment">## AAACATTGAGCTAC-1     pbmc3k       4903         1352  3.7935958</span><br><span class="hljs-comment">## AAACATTGATCAGC-1     pbmc3k       3147         1129  0.8897363</span><br><span class="hljs-comment">## AAACCGTGCTTCCG-1     pbmc3k       2639          960  1.7430845</span><br><span class="hljs-comment">## AAACCGTGTATGCG-1     pbmc3k        980          521  1.2244898</span><br></code></pre></td></tr></table></figure>

<p>可视化QC标准，<strong>quality control只是通过计算表示细胞质量，质控，但不筛选！</strong></p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><code class="hljs R"><span class="hljs-comment"># Visualize QC metrics as a violin plot</span><br>VlnPlot<span class="hljs-punctuation">(</span>pbmc<span class="hljs-punctuation">,</span> features <span class="hljs-operator">=</span> <span class="hljs-built_in">c</span><span class="hljs-punctuation">(</span><span class="hljs-string">&quot;nFeature_RNA&quot;</span><span class="hljs-punctuation">,</span> <span class="hljs-string">&quot;nCount_RNA&quot;</span><span class="hljs-punctuation">,</span> <span class="hljs-string">&quot;percent.mt&quot;</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">,</span> ncol <span class="hljs-operator">=</span> <span class="hljs-number">3</span><span class="hljs-punctuation">)</span><br></code></pre></td></tr></table></figure>

<p><img src="/../image/Seuratpictures%5Cqc2-1.png" srcset="/img/loading.gif" lazyload alt="qc2-1"></p>
<p>这里Count就是所有的RNA计数，Feature就是特征的RNA表达，这里的坐标轴不同，但是能看出nCountRNA分布更广泛，而mtRNA占比大多数都在2%附近。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><code class="hljs R"><span class="hljs-comment"># FeatureScatter is typically used to visualize feature-feature relationships, but can be used for anything calculated by the object, i.e. columns in object metadata, PC scores etc.</span><br><br>plot1 <span class="hljs-operator">&lt;-</span> FeatureScatter<span class="hljs-punctuation">(</span>pbmc<span class="hljs-punctuation">,</span> feature1 <span class="hljs-operator">=</span> <span class="hljs-string">&quot;nCount_RNA&quot;</span><span class="hljs-punctuation">,</span> feature2 <span class="hljs-operator">=</span> <span class="hljs-string">&quot;percent.mt&quot;</span><span class="hljs-punctuation">)</span><br>plot2 <span class="hljs-operator">&lt;-</span> FeatureScatter<span class="hljs-punctuation">(</span>pbmc<span class="hljs-punctuation">,</span> feature1 <span class="hljs-operator">=</span> <span class="hljs-string">&quot;nCount_RNA&quot;</span><span class="hljs-punctuation">,</span> feature2 <span class="hljs-operator">=</span> <span class="hljs-string">&quot;nFeature_RNA&quot;</span><span class="hljs-punctuation">)</span><br>plot1 <span class="hljs-operator">+</span> plot2<br></code></pre></td></tr></table></figure>

<p><img src="/../image/Seuratpictures%5Cqc2-2.png" srcset="/img/loading.gif" lazyload alt="qc2-2"></p>
<p>这两张图说明，特征RNA和总RNA是线性关系，但是mtRNA在总RNA多的时候往往少，总RNA少的时候往往多，不是线性关系，可能是总RNA多的时候往往细胞完整，RNA没有泄露，mtRNA占比少，总RNA少的时候往往细胞RNA丢失，线粒体中RNA保留，占比增加。<strong>图上方的是Pearson correlation</strong>。</p>
<blockquote>
<p>皮尔逊积矩相关系数（Pearson product-moment correlation coefficient， PPMCC），又叫样本相关系数。</p>
<p><img src="/../image/Seuratpictures%5C1.png" srcset="/img/loading.gif" lazyload alt="1"></p>
<p>Cov(X,Y)为X与Y的协方差，Var[X]为X的方差，Var[Y]为Y的方差。r的取值范围为-1≤r≤1，当r接近±1时表明观察的数据线性相关较强，当r接近0时表明观察数据无线性相关。当用样本相关系数来反映总体的变量之间是否相关，在样本容量比较小时通常需要进行相关系数的检验。</p>
</blockquote>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs R">pbmc <span class="hljs-operator">&lt;-</span> subset<span class="hljs-punctuation">(</span>pbmc<span class="hljs-punctuation">,</span> subset <span class="hljs-operator">=</span> nFeature_RNA <span class="hljs-operator">&gt;</span> <span class="hljs-number">200</span> <span class="hljs-operator">&amp;</span> nFeature_RNA <span class="hljs-operator">&lt;</span> <span class="hljs-number">2500</span> <span class="hljs-operator">&amp;</span> percent.mt <span class="hljs-operator">&lt;</span> <span class="hljs-number">5</span><span class="hljs-punctuation">)</span><br></code></pre></td></tr></table></figure>

<p>nFeature小于200说明质量不高，数量太少，可能是因为细胞破碎或死亡导致的RNA减少或者是空细胞槽，大于2500说明可能是多细胞，多细胞也不好，上文提到了。</p>
<h3 id="2-规范化细胞Normalizing-the-data"><a href="#2-规范化细胞Normalizing-the-data" class="headerlink" title="2.规范化细胞Normalizing the data"></a>2.规范化细胞Normalizing the data</h3><p>采用全局缩放归一化方法：LogNormalize。NormalizeData函数默认就是用LogNormalize。<br><strong>LogNormalize</strong>：对每个细胞的总表达量（total expression）乘以比例因子（scale factor），默认10000，然后取log。<strong>Normalized values are stored in .<code>pbmc[[&quot;RNA&quot;]]$data</code></strong></p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs R">pbmc <span class="hljs-operator">&lt;-</span> NormalizeData<span class="hljs-punctuation">(</span>pbmc<span class="hljs-punctuation">,</span> normalization.method <span class="hljs-operator">=</span> <span class="hljs-string">&quot;LogNormalize&quot;</span><span class="hljs-punctuation">,</span> scale.factor <span class="hljs-operator">=</span> <span class="hljs-number">10000</span><span class="hljs-punctuation">)</span><br></code></pre></td></tr></table></figure>

<p>全局扩展依赖于假设：每个细胞最初包含相同数量的总RNA。解决这个问题可以用ScTransform，后文补充。</p>
<h3 id="3-特征选择"><a href="#3-特征选择" class="headerlink" title="3.特征选择"></a>3.特征选择</h3><p>计算在数据集中表现出高细胞间变异的特征子集（即，它们在某些细胞中高表达，而在其他细胞中低表达），利用了<strong>标准差，越大说明变异越大</strong>。但是还不能完全理解在vst方法中要用到局部加权回归来取代方差。目前的理解：对于一种featureRNA，在有些细胞中高表达，有些细胞中低表达，还在大部分细胞中没有表达，局部方差回归想要找到存在于平均值和方差的关系，当平均值能够在一定程度上预测方差时，其比值更稳健。如果不用vst，用分箱，也是想说明标准差在不同的均值和方差的线性取值范围内，其关系不是恒定的，至少在分箱后不同箱内存在不同。通过局部加权回归和分箱，解决局部数据分布不均的问题。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><code class="hljs R">pbmc <span class="hljs-operator">&lt;-</span> FindVariableFeatures<span class="hljs-punctuation">(</span>pbmc<span class="hljs-punctuation">,</span> selection.method <span class="hljs-operator">=</span> <span class="hljs-string">&quot;vst&quot;</span><span class="hljs-punctuation">,</span> nfeatures <span class="hljs-operator">=</span> <span class="hljs-number">2000</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># Identify the 10 most highly variable genes</span><br>top10 <span class="hljs-operator">&lt;-</span> head<span class="hljs-punctuation">(</span>VariableFeatures<span class="hljs-punctuation">(</span>pbmc<span class="hljs-punctuation">)</span><span class="hljs-punctuation">,</span> <span class="hljs-number">10</span><span class="hljs-punctuation">)</span><br><br><span class="hljs-comment"># plot variable features with and without labels</span><br>plot1 <span class="hljs-operator">&lt;-</span> VariableFeaturePlot<span class="hljs-punctuation">(</span>pbmc<span class="hljs-punctuation">)</span><br>plot2 <span class="hljs-operator">&lt;-</span> LabelPoints<span class="hljs-punctuation">(</span>plot <span class="hljs-operator">=</span> plot1<span class="hljs-punctuation">,</span> points <span class="hljs-operator">=</span> top10<span class="hljs-punctuation">,</span> repel <span class="hljs-operator">=</span> <span class="hljs-literal">TRUE</span><span class="hljs-punctuation">)</span><br>plot1 <span class="hljs-operator">+</span> plot2<br></code></pre></td></tr></table></figure>

<p><img src="/../image/Seuratpictures%5Cvar_features-1.png" srcset="/img/loading.gif" lazyload alt="var_features-1"></p>
<p>选了2000个特征，然后标出了前十个。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br><span class="line">94</span><br><span class="line">95</span><br><span class="line">96</span><br><span class="line">97</span><br><span class="line">98</span><br><span class="line">99</span><br><span class="line">100</span><br><span class="line">101</span><br><span class="line">102</span><br><span class="line">103</span><br><span class="line">104</span><br><span class="line">105</span><br><span class="line">106</span><br><span class="line">107</span><br><span class="line">108</span><br><span class="line">109</span><br></pre></td><td class="code"><pre><code class="hljs R">FindVariableFeatures.V3Matrix <span class="hljs-operator">&lt;-</span> <span class="hljs-keyword">function</span><span class="hljs-punctuation">(</span><br>  object<span class="hljs-punctuation">,</span><br>  selection.method <span class="hljs-operator">=</span> <span class="hljs-string">&quot;vst&quot;</span><span class="hljs-punctuation">,</span> <span class="hljs-comment">#选择方法，vst或mvp</span><br>  loess.span <span class="hljs-operator">=</span> <span class="hljs-number">0.3</span><span class="hljs-punctuation">,</span> <span class="hljs-comment">#局部加权回归平滑度</span><br>  clip.max <span class="hljs-operator">=</span> <span class="hljs-string">&#x27;auto&#x27;</span><span class="hljs-punctuation">,</span> <span class="hljs-comment">#最大截断值</span><br>  mean.function <span class="hljs-operator">=</span> FastExpMean<span class="hljs-punctuation">,</span> <span class="hljs-comment">#均值计算</span><br>  dispersion.function <span class="hljs-operator">=</span> FastLogVMR<span class="hljs-punctuation">,</span> <span class="hljs-comment">#离散度计算</span><br>  num.bin <span class="hljs-operator">=</span> <span class="hljs-number">20</span><span class="hljs-punctuation">,</span> <span class="hljs-comment">#分箱数量</span><br>  binning.method <span class="hljs-operator">=</span> <span class="hljs-string">&quot;equal_width&quot;</span><span class="hljs-punctuation">,</span> <span class="hljs-comment">#分箱方法</span><br>  verbose <span class="hljs-operator">=</span> <span class="hljs-literal">TRUE</span><span class="hljs-punctuation">,</span> <span class="hljs-comment">#是否显示进度</span><br>  ...<br><span class="hljs-punctuation">)</span> <span class="hljs-punctuation">&#123;</span><br>  CheckDots<span class="hljs-punctuation">(</span>...<span class="hljs-punctuation">)</span><br>    <span class="hljs-comment">#不是矩阵的转为矩阵，不是压缩稀疏矩阵的转</span><br>  <span class="hljs-keyword">if</span> <span class="hljs-punctuation">(</span><span class="hljs-operator">!</span>inherits<span class="hljs-punctuation">(</span>x <span class="hljs-operator">=</span> object<span class="hljs-punctuation">,</span> <span class="hljs-string">&#x27;Matrix&#x27;</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span> <span class="hljs-punctuation">&#123;</span><br>    object <span class="hljs-operator">&lt;-</span> as<span class="hljs-punctuation">(</span>object <span class="hljs-operator">=</span> as.matrix<span class="hljs-punctuation">(</span>x <span class="hljs-operator">=</span> object<span class="hljs-punctuation">)</span><span class="hljs-punctuation">,</span> Class <span class="hljs-operator">=</span> <span class="hljs-string">&#x27;Matrix&#x27;</span><span class="hljs-punctuation">)</span><br>  <span class="hljs-punctuation">&#125;</span><br>  <span class="hljs-keyword">if</span> <span class="hljs-punctuation">(</span><span class="hljs-operator">!</span>inherits<span class="hljs-punctuation">(</span>x <span class="hljs-operator">=</span> object<span class="hljs-punctuation">,</span> what <span class="hljs-operator">=</span> <span class="hljs-string">&#x27;dgCMatrix&#x27;</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span> <span class="hljs-punctuation">&#123;</span><br>    object <span class="hljs-operator">&lt;-</span> as.sparse<span class="hljs-punctuation">(</span>x <span class="hljs-operator">=</span> object<span class="hljs-punctuation">)</span><br>  <span class="hljs-punctuation">&#125;</span><br>    <span class="hljs-comment">#检查方法是否为vst</span><br>  <span class="hljs-keyword">if</span> <span class="hljs-punctuation">(</span>selection.method <span class="hljs-operator">==</span> <span class="hljs-string">&quot;vst&quot;</span><span class="hljs-punctuation">)</span> <span class="hljs-punctuation">&#123;</span><br>    <span class="hljs-keyword">if</span> <span class="hljs-punctuation">(</span>clip.max <span class="hljs-operator">==</span> <span class="hljs-string">&#x27;auto&#x27;</span><span class="hljs-punctuation">)</span> <span class="hljs-punctuation">&#123;</span><br>        <span class="hljs-comment">#clip.max参数为auto，设置为数据对象列数的平方根</span><br>      clip.max <span class="hljs-operator">&lt;-</span> <span class="hljs-built_in">sqrt</span><span class="hljs-punctuation">(</span>x <span class="hljs-operator">=</span> ncol<span class="hljs-punctuation">(</span>x <span class="hljs-operator">=</span> object<span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span><br>    <span class="hljs-punctuation">&#125;</span><br>      <span class="hljs-comment">#创建一个包换行均值的数据框</span><br>    hvf.info <span class="hljs-operator">&lt;-</span> data.frame<span class="hljs-punctuation">(</span>mean <span class="hljs-operator">=</span> rowMeans<span class="hljs-punctuation">(</span>x <span class="hljs-operator">=</span> object<span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span><br>      <span class="hljs-comment">#SparseRowVar2函数的作用是针对稀疏矩阵中的每一行，以每行的均值信息作为输入，计算该行数据的方差</span><br>      hvf.info<span class="hljs-operator">$</span>variance <span class="hljs-operator">&lt;-</span> SparseRowVar2<span class="hljs-punctuation">(</span><br>      mat <span class="hljs-operator">=</span> object<span class="hljs-punctuation">,</span><br>      mu <span class="hljs-operator">=</span> hvf.info<span class="hljs-operator">$</span>mean<span class="hljs-punctuation">,</span><br>      display_progress <span class="hljs-operator">=</span> verbose<br>    <span class="hljs-punctuation">)</span><br>      <span class="hljs-comment">#初始化预期方差和标准化后的方差</span><br>    hvf.info<span class="hljs-operator">$</span>variance.expected <span class="hljs-operator">&lt;-</span> 0<br>    hvf.info<span class="hljs-operator">$</span>variance.standardized <span class="hljs-operator">&lt;-</span> 0<br>      <span class="hljs-comment">#创建逻辑向量，指示哪些行的方差大于0</span><br>    not.const <span class="hljs-operator">&lt;-</span> hvf.info<span class="hljs-operator">$</span>variance <span class="hljs-operator">&gt;</span> <span class="hljs-number">0</span><br>      <span class="hljs-comment">#局部加权回归（Locally Weighted Scatterplot Smoothing，Loess）是一种非参数的回归方法，用于拟合数据的非线性关系。与线性回归不同，Loess 使用局部加权的方式，对于每个数据点都赋予不同的权重，以适应数据的局部特征。</span><br>      <span class="hljs-comment"># Loess 的基本思想是对于每个数据点，利用附近的数据点进行局部的线性或多项式回归，然后通过加权平均的方式来获得最终的拟合曲线。这种加权平均使得拟合更加灵活，能够更好地适应数据的非线性关系。</span><br>      <span class="hljs-comment"># Loess 方法的一个重要参数是平滑度参数（span），它决定了用于拟合的局部数据点的数量，即局部拟合的程度。一般情况下，其取值范围通常在 0.1 到 0.5 之间。较小的平滑度参数会导致拟合曲线更接近原始数据，但也更容易受到噪声的影响；而较大的平滑度参数会导致更平滑的拟合曲线，但可能会忽略一些数据的局部特征</span><br>    fit <span class="hljs-operator">&lt;-</span> loess<span class="hljs-punctuation">(</span><br>        <span class="hljs-comment">#使用对数变换后的方差作为因变量，对数变换后的均值作为自变量进行拟合</span><br>      formula <span class="hljs-operator">=</span> log10<span class="hljs-punctuation">(</span>x <span class="hljs-operator">=</span> variance<span class="hljs-punctuation">)</span> <span class="hljs-operator">~</span> log10<span class="hljs-punctuation">(</span>x <span class="hljs-operator">=</span> mean<span class="hljs-punctuation">)</span><span class="hljs-punctuation">,</span><br>        <span class="hljs-comment">#用于筛选出方差大于零的数据点</span><br>      data <span class="hljs-operator">=</span> hvf.info<span class="hljs-punctuation">[</span>not.const<span class="hljs-punctuation">,</span> <span class="hljs-punctuation">]</span><span class="hljs-punctuation">,</span><br>        <span class="hljs-comment">#平滑度参数，控制了局部拟合的程度</span><br>      span <span class="hljs-operator">=</span> loess.span<br>    <span class="hljs-punctuation">)</span><br>      <span class="hljs-comment">#将局部加权回归（Loess）得到的拟合值赋给 hvf.info 数据框中的 variance.expected 列中方差大于0的行，这里用指数变换还原原始数据比例，loess中是用的log10(x=variance)</span><br>    hvf.info<span class="hljs-operator">$</span>variance.expected<span class="hljs-punctuation">[</span>not.const<span class="hljs-punctuation">]</span> <span class="hljs-operator">&lt;-</span> <span class="hljs-number">10</span> <span class="hljs-operator">^</span> fit<span class="hljs-operator">$</span>fitted<br>    <span class="hljs-comment"># use c function to get variance after feature standardization</span><br>    <span class="hljs-comment"># 计算稀疏矩阵每行的标准化后的方差（方差/期望（平均值）），消除尺度影响</span><br>    hvf.info<span class="hljs-operator">$</span>variance.standardized <span class="hljs-operator">&lt;-</span> SparseRowVarStd<span class="hljs-punctuation">(</span><br>      mat <span class="hljs-operator">=</span> object<span class="hljs-punctuation">,</span><br>        <span class="hljs-comment">#计算每行平均值</span><br>      mu <span class="hljs-operator">=</span> hvf.info<span class="hljs-operator">$</span>mean<span class="hljs-punctuation">,</span><br>        <span class="hljs-comment">#期望方差的平方根</span><br>      sd <span class="hljs-operator">=</span> <span class="hljs-built_in">sqrt</span><span class="hljs-punctuation">(</span>hvf.info<span class="hljs-operator">$</span>variance.expected<span class="hljs-punctuation">)</span><span class="hljs-punctuation">,</span><br>        <span class="hljs-comment">#控制方差上线，最大截断值</span><br>      vmax <span class="hljs-operator">=</span> clip.max<span class="hljs-punctuation">,</span><br>      display_progress <span class="hljs-operator">=</span> verbose<br>    <span class="hljs-punctuation">)</span><br>      <span class="hljs-comment">#使用 paste0 函数给每个列名添加前缀 vst.</span><br>    colnames<span class="hljs-punctuation">(</span>x <span class="hljs-operator">=</span> hvf.info<span class="hljs-punctuation">)</span> <span class="hljs-operator">&lt;-</span> paste0<span class="hljs-punctuation">(</span><span class="hljs-string">&#x27;vst.&#x27;</span><span class="hljs-punctuation">,</span> colnames<span class="hljs-punctuation">(</span>x <span class="hljs-operator">=</span> hvf.info<span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span><br>  <span class="hljs-punctuation">&#125;</span> <span class="hljs-keyword">else</span> <span class="hljs-punctuation">&#123;</span><br>    <span class="hljs-keyword">if</span> <span class="hljs-punctuation">(</span><span class="hljs-operator">!</span>inherits<span class="hljs-punctuation">(</span>x <span class="hljs-operator">=</span> mean.function<span class="hljs-punctuation">,</span> what <span class="hljs-operator">=</span> <span class="hljs-string">&#x27;function&#x27;</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span> <span class="hljs-punctuation">&#123;</span><br>      stop<span class="hljs-punctuation">(</span><span class="hljs-string">&quot;&#x27;mean.function&#x27; must be a function&quot;</span><span class="hljs-punctuation">)</span><br>    <span class="hljs-punctuation">&#125;</span><br>    <span class="hljs-keyword">if</span> <span class="hljs-punctuation">(</span><span class="hljs-operator">!</span>inherits<span class="hljs-punctuation">(</span>x <span class="hljs-operator">=</span> dispersion.function<span class="hljs-punctuation">,</span> what <span class="hljs-operator">=</span> <span class="hljs-string">&#x27;function&#x27;</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span> <span class="hljs-punctuation">&#123;</span><br>      stop<span class="hljs-punctuation">(</span><span class="hljs-string">&quot;&#x27;dispersion.function&#x27; must be a function&quot;</span><span class="hljs-punctuation">)</span><br>    <span class="hljs-punctuation">&#125;</span><br>    feature.mean <span class="hljs-operator">&lt;-</span> mean.function<span class="hljs-punctuation">(</span>object<span class="hljs-punctuation">,</span> verbose<span class="hljs-punctuation">)</span><br>    feature.dispersion <span class="hljs-operator">&lt;-</span> dispersion.function<span class="hljs-punctuation">(</span>object<span class="hljs-punctuation">,</span> verbose<span class="hljs-punctuation">)</span><br>      <span class="hljs-comment">#链式赋值</span><br>    <span class="hljs-built_in">names</span><span class="hljs-punctuation">(</span>x <span class="hljs-operator">=</span> feature.mean<span class="hljs-punctuation">)</span> <span class="hljs-operator">&lt;-</span> <span class="hljs-built_in">names</span><span class="hljs-punctuation">(</span>x <span class="hljs-operator">=</span> feature.dispersion<span class="hljs-punctuation">)</span> <span class="hljs-operator">&lt;-</span> rownames<span class="hljs-punctuation">(</span>x <span class="hljs-operator">=</span> object<span class="hljs-punctuation">)</span><br>      <span class="hljs-comment">#将缺失值替换为0</span><br>    feature.dispersion<span class="hljs-punctuation">[</span><span class="hljs-built_in">is.na</span><span class="hljs-punctuation">(</span>x <span class="hljs-operator">=</span> feature.dispersion<span class="hljs-punctuation">)</span><span class="hljs-punctuation">]</span> <span class="hljs-operator">&lt;-</span> <span class="hljs-number">0</span><br>    feature.mean<span class="hljs-punctuation">[</span><span class="hljs-built_in">is.na</span><span class="hljs-punctuation">(</span>x <span class="hljs-operator">=</span> feature.mean<span class="hljs-punctuation">)</span><span class="hljs-punctuation">]</span> <span class="hljs-operator">&lt;-</span> <span class="hljs-number">0</span><br>      <span class="hljs-comment">#条件语句switch根据binning.method的取值选择不同的分箱方法，这里不是equal_width，就是equal_frequency</span><br>    data.x.breaks <span class="hljs-operator">&lt;-</span> <span class="hljs-built_in">switch</span><span class="hljs-punctuation">(</span><br>      EXPR <span class="hljs-operator">=</span> binning.method<span class="hljs-punctuation">,</span><br>      <span class="hljs-string">&#x27;equal_width&#x27;</span> <span class="hljs-operator">=</span> num.bin<span class="hljs-punctuation">,</span><br>      <span class="hljs-string">&#x27;equal_frequency&#x27;</span> <span class="hljs-operator">=</span> <span class="hljs-built_in">c</span><span class="hljs-punctuation">(</span><br>        <span class="hljs-operator">-</span><span class="hljs-number">1</span><span class="hljs-punctuation">,</span><br>          <span class="hljs-comment">#利用 quantile 函数计算等频率分箱的分位数</span><br>        quantile<span class="hljs-punctuation">(</span><br>          x <span class="hljs-operator">=</span> feature.mean<span class="hljs-punctuation">[</span>feature.mean <span class="hljs-operator">&gt;</span> <span class="hljs-number">0</span><span class="hljs-punctuation">]</span><span class="hljs-punctuation">,</span><br>            <span class="hljs-comment">#seq.int从0到1生成等间隔分位数</span><br>          probs <span class="hljs-operator">=</span> <span class="hljs-built_in">seq.int</span><span class="hljs-punctuation">(</span>from <span class="hljs-operator">=</span> <span class="hljs-number">0</span><span class="hljs-punctuation">,</span> to <span class="hljs-operator">=</span> <span class="hljs-number">1</span><span class="hljs-punctuation">,</span> length.out <span class="hljs-operator">=</span> num.bin<span class="hljs-punctuation">)</span><br>        <span class="hljs-punctuation">)</span><br>      <span class="hljs-punctuation">)</span><span class="hljs-punctuation">,</span><br>      stop<span class="hljs-punctuation">(</span><span class="hljs-string">&quot;Unknown binning method: &quot;</span><span class="hljs-punctuation">,</span> binning.method<span class="hljs-punctuation">)</span><br>    <span class="hljs-punctuation">)</span><br>    data.x.bin <span class="hljs-operator">&lt;-</span> cut<span class="hljs-punctuation">(</span>x <span class="hljs-operator">=</span> feature.mean<span class="hljs-punctuation">,</span> breaks <span class="hljs-operator">=</span> data.x.breaks<span class="hljs-punctuation">)</span><br>    <span class="hljs-built_in">names</span><span class="hljs-punctuation">(</span>x <span class="hljs-operator">=</span> data.x.bin<span class="hljs-punctuation">)</span> <span class="hljs-operator">&lt;-</span> <span class="hljs-built_in">names</span><span class="hljs-punctuation">(</span>x <span class="hljs-operator">=</span> feature.mean<span class="hljs-punctuation">)</span><br>    mean.y <span class="hljs-operator">&lt;-</span> tapply<span class="hljs-punctuation">(</span>X <span class="hljs-operator">=</span> feature.dispersion<span class="hljs-punctuation">,</span> INDEX <span class="hljs-operator">=</span> data.x.bin<span class="hljs-punctuation">,</span> FUN <span class="hljs-operator">=</span> mean<span class="hljs-punctuation">)</span><br>    sd.y <span class="hljs-operator">&lt;-</span> tapply<span class="hljs-punctuation">(</span>X <span class="hljs-operator">=</span> feature.dispersion<span class="hljs-punctuation">,</span> INDEX <span class="hljs-operator">=</span> data.x.bin<span class="hljs-punctuation">,</span> FUN <span class="hljs-operator">=</span> sd<span class="hljs-punctuation">)</span><br>      <span class="hljs-comment">#每个特征的离散度减去对应分箱的均值，并除以标准差。</span><br>    feature.dispersion.scaled <span class="hljs-operator">&lt;-</span> <span class="hljs-punctuation">(</span>feature.dispersion <span class="hljs-operator">-</span> mean.y<span class="hljs-punctuation">[</span><span class="hljs-built_in">as.numeric</span><span class="hljs-punctuation">(</span>x <span class="hljs-operator">=</span> data.x.bin<span class="hljs-punctuation">)</span><span class="hljs-punctuation">]</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">/</span><br>      sd.y<span class="hljs-punctuation">[</span><span class="hljs-built_in">as.numeric</span><span class="hljs-punctuation">(</span>x <span class="hljs-operator">=</span> data.x.bin<span class="hljs-punctuation">)</span><span class="hljs-punctuation">]</span><br>    <span class="hljs-built_in">names</span><span class="hljs-punctuation">(</span>x <span class="hljs-operator">=</span> feature.dispersion.scaled<span class="hljs-punctuation">)</span> <span class="hljs-operator">&lt;-</span> <span class="hljs-built_in">names</span><span class="hljs-punctuation">(</span>x <span class="hljs-operator">=</span> feature.mean<span class="hljs-punctuation">)</span><br>    hvf.info <span class="hljs-operator">&lt;-</span> data.frame<span class="hljs-punctuation">(</span>feature.mean<span class="hljs-punctuation">,</span> feature.dispersion<span class="hljs-punctuation">,</span> feature.dispersion.scaled<span class="hljs-punctuation">)</span><br>    rownames<span class="hljs-punctuation">(</span>x <span class="hljs-operator">=</span> hvf.info<span class="hljs-punctuation">)</span> <span class="hljs-operator">&lt;-</span> rownames<span class="hljs-punctuation">(</span>x <span class="hljs-operator">=</span> object<span class="hljs-punctuation">)</span><br>    colnames<span class="hljs-punctuation">(</span>x <span class="hljs-operator">=</span> hvf.info<span class="hljs-punctuation">)</span> <span class="hljs-operator">&lt;-</span> paste0<span class="hljs-punctuation">(</span><span class="hljs-string">&#x27;mvp.&#x27;</span><span class="hljs-punctuation">,</span> <span class="hljs-built_in">c</span><span class="hljs-punctuation">(</span><span class="hljs-string">&#x27;mean&#x27;</span><span class="hljs-punctuation">,</span> <span class="hljs-string">&#x27;dispersion&#x27;</span><span class="hljs-punctuation">,</span> <span class="hljs-string">&#x27;dispersion.scaled&#x27;</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span><br>  <span class="hljs-punctuation">&#125;</span><br>  <span class="hljs-built_in">return</span><span class="hljs-punctuation">(</span>hvf.info<span class="hljs-punctuation">)</span><br><span class="hljs-punctuation">&#125;</span><br></code></pre></td></tr></table></figure>

<h3 id="4-缩放数据Scaling-data"><a href="#4-缩放数据Scaling-data" class="headerlink" title="4.缩放数据Scaling data"></a>4.缩放数据Scaling data</h3><p>是PCA等降维技术之前的预处理步骤，前面的LogNormalize是对列和列之间处理，让每列的细胞的表达RNA数量相同，基于前提假设。现在是对每行进行处理，让每行的RNA在所有细胞中的表达进行缩放。每行基因的在细胞中表达平均为0，方差为1，平均为0就是对表达水平进行标准化，方差为1是对离散程度进行标准化。这一步为了做PCA时，确保每个特征对主成分的贡献相对一致。</p>
<figure class="highlight scss"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><code class="hljs scss"><span class="hljs-attribute">all</span><span class="hljs-selector-class">.genes</span> &lt;- <span class="hljs-built_in">rownames</span>(pbmc)<br>pbmc &lt;- <span class="hljs-built_in">ScaleData</span>(pbmc, features = all.genes)<br></code></pre></td></tr></table></figure>

<p>这里挖个坑，如果要去除细胞周期影响和mtRNA影响，最好用sctransform</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs R">pbmc <span class="hljs-operator">&lt;-</span> ScaleData<span class="hljs-punctuation">(</span>pbmc<span class="hljs-punctuation">,</span> vars.to.regress <span class="hljs-operator">=</span> <span class="hljs-string">&quot;percent.mt&quot;</span><span class="hljs-punctuation">)</span><br></code></pre></td></tr></table></figure>

<h2 id="二、线性降维"><a href="#二、线性降维" class="headerlink" title="二、线性降维"></a>二、线性降维</h2><blockquote>
<p>PCA 是主成分分析（Principal Component Analysis）的缩写。它是一种常用的<strong>数据降维</strong>技术，用于找到数据集中最重要的特征，并将数据映射到一个新的坐标系上，以便更好地理解数据的结构和模式。</p>
<p>在 PCA 中，通过线性变换将原始的高维数据转换成一组低维的特征，这些特征被称为主成分。这些主成分是原始特征的<strong>线性组合</strong>，按照重要性依次排列，<strong>第一个主成分包含最大方差</strong>，第二个主成分包含次大方差，以此类推。通过保留最重要的主成分，可以实现数据的降维，同时尽可能地保留原始数据的信息。</p>
</blockquote>
<p>来自<em>解决数据挖掘文章的figure1</em>的群友“池塘春草梦” 的回答：</p>
<blockquote>
<p>要弄清楚positive和negative的意思，首先要知道pca是什么。</p>
<p>pca的全称是Principal Component Analysis，三个单词分别就是主 成分 分析。</p>
<p>为什么要做pca？首先看看我们的输入，上面代码的输入应该是cell*gene矩阵，一般来说这个基因个数很多，或成为“特征”很多，但这为我们的分析带来了麻烦。</p>
<p>比如说我们想区分人的性别，于是我们开始搜集人的各种特征（头发长度、眼睛个数、鼻子个数、喉结个数……）。我们会发现：如果我们考虑所有的特征，分类效果很一般，因为有很多特征都是和性别无关的，比如眼睛和鼻子个数。</p>
<p>那该怎么办？一个思路就是我把这些特征融合起来，比如把眼睛鼻子个数融合起来成为一个新的特征、头发长度和喉结个数融合……</p>
<p>这些东西也就是所谓的主成分，通俗理解就是这些特征的主要成分。</p>
<p>专业一点就是pca通过正交变换将原始特征空间中的线性相关变量转换为新的线性无关变量，也就是主成分。</p>
<p>这些主成分的数量少于或等于原始特征的数量，同时能够保留原始数据的大部分变异性。</p>
<p>………分割线……………</p>
<p>了解了pca的原理，我们再来看看单细胞中pca的结果，刚刚我们输入cell*gene的矩阵，进行pca降维，用少数几个主成分表示了所有基因。</p>
<p>每个主成分（pc）通常会有一系列与之相关的系数（或称为载荷），这些系数有正有负。这些系数的正负和大小可以帮助我们理解每个主成分与原始特征的关系。</p>
<p>Positive（正）: 如果某个特征在主成分上的载荷为正，那么这意味着这个特征与主成分之间存在正相关关系。换句话说，当这个特征的值增加时，主成分的值也倾向于增加。<br>Negative（负）: 如果载荷为负，则表示该特征与主成分之间存在负相关关系。当这个特征的值增加时，主成分的值会倾向于减少。</p>
<p>在你给出的第1个图中，他对每个主成分都给出了相应的正相关基因和负相关基因。</p>
<p>第2个图就是把每个主成分（这里是pc1和pc2）中按照载荷从大到小排列。</p>
<p>横坐标就是指的载荷的大小，越大就是越正相关，越负就是越负相关。</p>
<p>纵坐标就是基因了，可以看到图2和图1的结果，都是能对应上的。</p>
</blockquote>
<p>当你进行 PCA 时，你会按照以下步骤进行详细的计算：</p>
<ol>
<li><p><strong>数据标准化</strong></p>
</li>
<li><p><strong>计算协方差矩阵</strong></p>
</li>
<li><p><strong>特征值分解</strong></p>
</li>
<li><p><strong>选择主成分</strong></p>
</li>
<li><p><strong>数据转换</strong></p>
</li>
</ol>
<p><img src="/../image/Seuratpictures%5CPCA1.png" srcset="/img/loading.gif" lazyload alt="PCA1"></p>
<p><img src="/../image/Seuratpictures%5CPCA2.png" srcset="/img/loading.gif" lazyload alt="PCA2"></p>
<p>通过这些步骤，你可以得到一个降维后的数据集，其中每个样本都被表示为一组新的主成分。这些主成分捕捉了原始数据中最重要的结构和变化。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs R">pbmc <span class="hljs-operator">&lt;-</span> RunPCA<span class="hljs-punctuation">(</span>pbmc<span class="hljs-punctuation">,</span> features <span class="hljs-operator">=</span> VariableFeatures<span class="hljs-punctuation">(</span>object <span class="hljs-operator">=</span> pbmc<span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span><br></code></pre></td></tr></table></figure>

<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br></pre></td><td class="code"><pre><code class="hljs R"><span class="hljs-comment"># Examine and visualize PCA results a few different ways</span><br><span class="hljs-comment"># pbmc[[&quot;pca&quot;]] 表示从 pbmc 对象中获取 PCA 分析的结果</span><br><span class="hljs-comment"># 使用 dims = 1:5 参数指定打印前 5 个主成分的信息</span><br><span class="hljs-comment"># nfeatures = 5 参数指定每个主成分的前 5 个特征（基因）的信息。</span><br>print<span class="hljs-punctuation">(</span>pbmc<span class="hljs-punctuation">[[</span><span class="hljs-string">&quot;pca&quot;</span><span class="hljs-punctuation">]</span><span class="hljs-punctuation">]</span><span class="hljs-punctuation">,</span> dims <span class="hljs-operator">=</span> <span class="hljs-number">1</span><span class="hljs-operator">:</span><span class="hljs-number">5</span><span class="hljs-punctuation">,</span> nfeatures <span class="hljs-operator">=</span> <span class="hljs-number">5</span><span class="hljs-punctuation">)</span><br><span class="hljs-comment">## PC_ 1 </span><br><span class="hljs-comment">## Positive:  CST3, TYROBP, LST1, AIF1, FTL </span><br><span class="hljs-comment">## Negative:  MALAT1, LTB, IL32, IL7R, CD2 </span><br><span class="hljs-comment">## PC_ 2 </span><br><span class="hljs-comment">## Positive:  CD79A, MS4A1, TCL1A, HLA-DQA1, HLA-DQB1 </span><br><span class="hljs-comment">## Negative:  NKG7, PRF1, CST7, GZMB, GZMA </span><br><span class="hljs-comment">## PC_ 3 </span><br><span class="hljs-comment">## Positive:  HLA-DQA1, CD79A, CD79B, HLA-DQB1, HLA-DPB1 </span><br><span class="hljs-comment">## Negative:  PPBP, PF4, SDPR, SPARC, GNG11 </span><br><span class="hljs-comment">## PC_ 4 </span><br><span class="hljs-comment">## Positive:  HLA-DQA1, CD79B, CD79A, MS4A1, HLA-DQB1 </span><br><span class="hljs-comment">## Negative:  VIM, IL7R, S100A6, IL32, S100A8 </span><br><span class="hljs-comment">## PC_ 5 </span><br><span class="hljs-comment">## Positive:  GZMB, NKG7, S100A8, FGFBP2, GNLY </span><br><span class="hljs-comment">## Negative:  LTB, IL7R, CKB, VIM, MS4A7</span><br></code></pre></td></tr></table></figure>

<blockquote>
<p>在 R 中，使用单括号 <code>[ ]</code> 用于提取列表对象中的元素，而使用双括号 <code>[[ ]]</code> 用于提取列表对象中的元素值。这两种方法都可以用于提取列表中的元素，但在不同的情况下有不同的用途：</p>
<ul>
<li>单括号 <code>[ ]</code> 用于提取子集，返回的是一个列表或者一个子集的列表。</li>
<li>双括号 <code>[[ ]]</code> 用于提取元素的值，返回的是元素的实际值。</li>
</ul>
<p>对于 Seurat 对象中存储的 PCA 结果，**<code>pbmc[[&quot;pca&quot;]]</code> 用于获取 PCA 结果的值，而 <code>pbmc[&quot;pca&quot;]</code> 则用于获取包含 PCA 结果的列表**。在这种情况下，我们通常想要获取 PCA 结果的值，因此使用双括号来访问这个元素的值是更常见的做法。</p>
</blockquote>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs R">VizDimLoadings<span class="hljs-punctuation">(</span>pbmc<span class="hljs-punctuation">,</span> dims <span class="hljs-operator">=</span> <span class="hljs-number">1</span><span class="hljs-operator">:</span><span class="hljs-number">2</span><span class="hljs-punctuation">,</span> reduction <span class="hljs-operator">=</span> <span class="hljs-string">&quot;pca&quot;</span><span class="hljs-punctuation">)</span><br></code></pre></td></tr></table></figure>

<p>左图是第一个主成分的信息，横轴应该是主成分得分，正表示正相关、负表示负相关。</p>
<p><img src="/../image/Seuratpictures%5Cpca_viz-1.png" srcset="/img/loading.gif" lazyload alt="pca_viz-1"></p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs R">DimPlot<span class="hljs-punctuation">(</span>pbmc<span class="hljs-punctuation">,</span> reduction <span class="hljs-operator">=</span> <span class="hljs-string">&quot;pca&quot;</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span> NoLegend<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><br></code></pre></td></tr></table></figure>

<p><img src="/../image/Seuratpictures%5Cpca_viz-2.png" srcset="/img/loading.gif" lazyload alt="pca_viz-2"></p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs R">DimHeatmap<span class="hljs-punctuation">(</span>pbmc<span class="hljs-punctuation">,</span> dims <span class="hljs-operator">=</span> <span class="hljs-number">1</span><span class="hljs-punctuation">,</span> cells <span class="hljs-operator">=</span> <span class="hljs-number">500</span><span class="hljs-punctuation">,</span> balanced <span class="hljs-operator">=</span> <span class="hljs-literal">TRUE</span><span class="hljs-punctuation">)</span><br></code></pre></td></tr></table></figure>

<p><img src="/../image/Seuratpictures%5Csingle-heatmap-1.png" srcset="/img/loading.gif" lazyload alt="single-heatmap-1"></p>
<p>PCA可以探索数据集中异质性的主要来源，通过画PCA热图看。</p>
<p>这一步是没有返回值的，不过作者改过代码，我装的版本还没改，所以要加上fast&#x3D;FALSE</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs R">DimHeatmap<span class="hljs-punctuation">(</span>pbmc<span class="hljs-punctuation">,</span> dims <span class="hljs-operator">=</span> <span class="hljs-number">1</span><span class="hljs-operator">:</span><span class="hljs-number">15</span><span class="hljs-punctuation">,</span> cells <span class="hljs-operator">=</span> <span class="hljs-number">500</span><span class="hljs-punctuation">,</span> balanced <span class="hljs-operator">=</span> <span class="hljs-literal">TRUE</span><span class="hljs-punctuation">,</span> fast <span class="hljs-operator">=</span> <span class="hljs-literal">FALSE</span><span class="hljs-punctuation">)</span><br></code></pre></td></tr></table></figure>

<p>![multi-heatmap-1 (1)](..&#x2F;image&#x2F;Seuratpictures\multi-heatmap-1 (1).png)</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs R">ElbowPlot<span class="hljs-punctuation">(</span>pbmc<span class="hljs-punctuation">)</span><br></code></pre></td></tr></table></figure>

<p>对每一个主成分可解释的百分比进行排序</p>
<p><img src="/../image/Seuratpictures%5Celbow_plot-1.png" srcset="/img/loading.gif" lazyload alt="elbow_plot-1"></p>
<p>在选择时如果主成分包括重要的需要被研究考虑的基因，应予以保留。可以尝试保留不同数量的主成分进行分析（10,15,50），但是结果不会大变。不要选择主成分时太苛刻，保留太少如5个会对下游分析造成不好的影响。</p>
<h2 id="三、细胞聚类"><a href="#三、细胞聚类" class="headerlink" title="三、细胞聚类"></a>三、细胞聚类</h2><p>采用基于图的聚类，K近邻法（k-nearset neighbor, KNN）</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><code class="hljs R"><span class="hljs-comment">#FindNeighbors，dims是在PCA后自行选择的维度，计算任意两个细胞之间的权重</span><br>pbmc <span class="hljs-operator">&lt;-</span> FindNeighbors<span class="hljs-punctuation">(</span>pbmc<span class="hljs-punctuation">,</span> dims <span class="hljs-operator">=</span> <span class="hljs-number">1</span><span class="hljs-operator">:</span><span class="hljs-number">10</span><span class="hljs-punctuation">)</span><br><span class="hljs-comment"># 聚类，分辨率参数，值越大，聚类数量越多。</span><br><span class="hljs-comment"># 0.4-1.2 之间设置此参数通常对于大约3000个细胞的单细胞数据集返回良好的结果。对于较大的数据集，最佳分辨率通常会提高</span><br>pbmc <span class="hljs-operator">&lt;-</span> FindClusters<span class="hljs-punctuation">(</span>pbmc<span class="hljs-punctuation">,</span> resolution <span class="hljs-operator">=</span> <span class="hljs-number">0.5</span><span class="hljs-punctuation">)</span><br></code></pre></td></tr></table></figure>

<p>这里resolutions对于大的数据集分辨率会提高，但是如果是小的数据集，大了以后在umap图上聚类会重叠，我直观上认为这样不好，感觉像是机器学习中的过度拟合一样。</p>
<h2 id="四、非线性降维（UMAP-tSNE）"><a href="#四、非线性降维（UMAP-tSNE）" class="headerlink" title="四、非线性降维（UMAP&#x2F;tSNE）"></a>四、非线性降维（UMAP&#x2F;tSNE）</h2><p>tSNE比UMAP聚类能分得更开，但是图里的细胞距离没有意义，UMAP图在距离里近的细胞相关性强。</p>
<p>To learn underlying structure in the dataset, in order to place similar cells together in low-dimensional space。所以基于图的聚类后相似的细胞应该定位在相同的位置。所以如果分辨率太高是没有意义的。而且本身这种降维聚类后再可视化有很大的局限性，聚类损失信息，可视化更损失信息，而且带有观察者的主观判断。而且这样的只会考虑细胞间的局部距离而忽略了全局定位。要避免仅仅基于可视化技术得出生物结论。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><code class="hljs R">pbmc <span class="hljs-operator">&lt;-</span> RunUMAP<span class="hljs-punctuation">(</span>pbmc<span class="hljs-punctuation">,</span> dims <span class="hljs-operator">=</span> <span class="hljs-number">1</span><span class="hljs-operator">:</span><span class="hljs-number">10</span><span class="hljs-punctuation">)</span><br><span class="hljs-comment"># note that you can set `label = TRUE` or use the LabelClusters function to help label</span><br><span class="hljs-comment"># individual clusters</span><br>DimPlot<span class="hljs-punctuation">(</span>pbmc<span class="hljs-punctuation">,</span> reduction <span class="hljs-operator">=</span> <span class="hljs-string">&quot;umap&quot;</span><span class="hljs-punctuation">)</span><br>saveRDS<span class="hljs-punctuation">(</span>pbmc<span class="hljs-punctuation">,</span> file <span class="hljs-operator">=</span> <span class="hljs-string">&quot;../output/pbmc_tutorial.rds&quot;</span><span class="hljs-punctuation">)</span><br></code></pre></td></tr></table></figure>

<p><strong>关于保存rds</strong>：</p>
<p>可以不需要重新运行scaledata，findneighbor，findcluster，runumap等计算步骤了，但是保存的时候会比较久，rds文件较大，几个G，需要预留时间空间。而且只要保存一次，就要把saveRDS注释掉，否则有可能在新的工作环境中运行了saveRDS，用空白的rds覆盖了原先的。</p>
<p><img src="/../image/Seuratpictures%5Cumapplot-1.png" srcset="/img/loading.gif" lazyload alt="umapplot-1"></p>
<p><img src="/../image/Seuratpictures%5Cumap-resolution=2.png" srcset="/img/loading.gif" lazyload alt="umap-resolution=2"></p>
<p>上面这个聚类resolution为2，下面为0.8，在有些地方似乎分开更清晰，但更大分辨率此处明显是不必要的。</p>
<p><img src="/../image/Seuratpictures%5Cumap-resolution=0.8.png" srcset="/img/loading.gif" lazyload alt="umap-resolution=0.8"></p>
<h2 id="五、寻找聚类marker（差异表达的特征基因）"><a href="#五、寻找聚类marker（差异表达的特征基因）" class="headerlink" title="五、寻找聚类marker（差异表达的特征基因）"></a>五、寻找聚类marker（差异表达的特征基因）</h2><p>寻找是哪些差异表达的基因决定了聚类结果，也就是找聚类marker（标记基因）。将每个细胞与其他细胞相比，找出<strong>正相关与负相关</strong>的（positive，negative）基因。这是通过FindAllMarkers函数实现的。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><code class="hljs R"><span class="hljs-comment"># find all markers of cluster 2</span><br>cluster2.markers <span class="hljs-operator">&lt;-</span> FindMarkers<span class="hljs-punctuation">(</span>pbmc<span class="hljs-punctuation">,</span> ident.1 <span class="hljs-operator">=</span> <span class="hljs-number">2</span><span class="hljs-punctuation">)</span><br>head<span class="hljs-punctuation">(</span>cluster2.markers<span class="hljs-punctuation">,</span> n <span class="hljs-operator">=</span> <span class="hljs-number">5</span><span class="hljs-punctuation">)</span><br></code></pre></td></tr></table></figure>

<p>FindMarkers用到了很多统计方法，在这挖个坑。</p>
<blockquote>
<p>看R包内函数的说明文档：</p>
<p>在R中，<code>#&#39;</code> 是一种用于注释的特殊标记，通常用于文档注释和函数文档字符串中。在这里，<code>#&#39;</code> 用于注释函数参数的描述。参数描述通常包含在 <code>@param</code> 标签后面。</p>
<p><code>#&#39; @param</code> 表示一个函数参数的描述，后面跟着参数的名称。<code>#&#39; \item&#123;&#125;</code> 用于列出参数的选项或说明。在这种情况下，<code>\item&#123;&#125;</code> 是在 <code>@param</code> 内部用于描述参数可能的选项或值。</p>
<p>这种表示格式通常被称为”<strong>Roxygen</strong>“格式，用于R语言中的文档注释。Roxygen注释结合了R代码和相关的文档信息，可以通过Roxygen软件包生成文档。一旦文档被正确注释，可以使用Roxygen软件包中的功能自动生成文档，这些文档通常被称为”<strong>Rd文件</strong>“，并且可以被R中的<code>?</code>和<code>help()</code>函数调用，以及其他一些工具和IDE使用。</p>
<p>在GitHub中，这种注释格式确实也很方便，因为GitHub支持Markdown格式，而Roxygen的注释风格很接近Markdown风格，因此可以直接在GitHub上清晰地显示文档注释。</p>
</blockquote>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br></pre></td><td class="code"><pre><code class="hljs R"><span class="hljs-comment">#&#x27; <span class="hljs-doctag">@param <span class="hljs-variable">test.use</span></span> Denotes which test to use. Available options are:</span><br><span class="hljs-comment">#&#x27; <span class="hljs-keyword">\itemize</span>&#123;</span><br><span class="hljs-comment">#&#x27;  <span class="hljs-keyword">\item</span>&#123;&quot;wilcox&quot;&#125; : Identifies differentially expressed genes between two</span><br><span class="hljs-comment">#&#x27;  groups of cells using a Wilcoxon Rank Sum test (default); will use a fast</span><br><span class="hljs-comment">#&#x27;  implementation by Presto if installed</span><br><span class="hljs-comment">#&#x27;  <span class="hljs-keyword">\item</span>&#123;&quot;wilcox_limma&quot;&#125; : Identifies differentially expressed genes between two</span><br><span class="hljs-comment">#&#x27;  groups of cells using the limma implementation of the Wilcoxon Rank Sum test;</span><br><span class="hljs-comment">#&#x27;  set this option to reproduce results from Seurat v4</span><br><span class="hljs-comment">#&#x27;  <span class="hljs-keyword">\item</span>&#123;&quot;bimod&quot;&#125; : Likelihood-ratio test for single cell gene expression,</span><br><span class="hljs-comment">#&#x27;  (McDavid et al., Bioinformatics, 2013)</span><br><span class="hljs-comment">#&#x27;  <span class="hljs-keyword">\item</span>&#123;&quot;roc&quot;&#125; : Identifies &#x27;markers&#x27; of gene expression using ROC analysis.</span><br><span class="hljs-comment">#&#x27;  For each gene, evaluates (using AUC) a classifier built on that gene alone,</span><br><span class="hljs-comment">#&#x27;  to classify between two groups of cells. An AUC value of 1 means that</span><br><span class="hljs-comment">#&#x27;  expression values for this gene alone can perfectly classify the two</span><br><span class="hljs-comment">#&#x27;  groupings (i.e. Each of the cells in cells.1 exhibit a higher level than</span><br><span class="hljs-comment">#&#x27;  each of the cells in cells.2). An AUC value of 0 also means there is perfect</span><br><span class="hljs-comment">#&#x27;  classification, but in the other direction. A value of 0.5 implies that</span><br><span class="hljs-comment">#&#x27;  the gene has no predictive power to classify the two groups. Returns a</span><br><span class="hljs-comment">#&#x27;  &#x27;predictive power&#x27; (abs(AUC-0.5) * 2) ranked matrix of putative differentially</span><br><span class="hljs-comment">#&#x27;  expressed genes.</span><br><span class="hljs-comment">#&#x27;  <span class="hljs-keyword">\item</span>&#123;&quot;t&quot;&#125; : Identify differentially expressed genes between two groups of</span><br><span class="hljs-comment">#&#x27;  cells using the Student&#x27;s t-test.</span><br><span class="hljs-comment">#&#x27;  <span class="hljs-keyword">\item</span>&#123;&quot;negbinom&quot;&#125; : Identifies differentially expressed genes between two</span><br><span class="hljs-comment">#&#x27;   groups of cells using a negative binomial generalized linear model.</span><br><span class="hljs-comment">#&#x27;   Use only for UMI-based datasets</span><br><span class="hljs-comment">#&#x27;  <span class="hljs-keyword">\item</span>&#123;&quot;poisson&quot;&#125; : Identifies differentially expressed genes between two</span><br><span class="hljs-comment">#&#x27;   groups of cells using a poisson generalized linear model.</span><br><span class="hljs-comment">#&#x27;   Use only for UMI-based datasets</span><br><span class="hljs-comment">#&#x27;  <span class="hljs-keyword">\item</span>&#123;&quot;LR&quot;&#125; : Uses a logistic regression framework to determine differentially</span><br><span class="hljs-comment">#&#x27;  expressed genes. Constructs a logistic regression model predicting group</span><br><span class="hljs-comment">#&#x27;  membership based on each feature individually and compares this to a null</span><br><span class="hljs-comment">#&#x27;  model with a likelihood ratio test.</span><br><span class="hljs-comment">#&#x27;  <span class="hljs-keyword">\item</span>&#123;&quot;MAST&quot;&#125; : Identifies differentially expressed genes between two groups</span><br><span class="hljs-comment">#&#x27;  of cells using a hurdle model tailored to scRNA-seq data. Utilizes the MAST</span><br><span class="hljs-comment">#&#x27;  package to run the DE testing.</span><br><span class="hljs-comment">#&#x27;  <span class="hljs-keyword">\item</span>&#123;&quot;DESeq2&quot;&#125; : Identifies differentially expressed genes between two groups</span><br><span class="hljs-comment">#&#x27;  of cells based on a model using DESeq2 which uses a negative binomial</span><br><span class="hljs-comment">#&#x27;  distribution (Love et al, Genome Biology, 2014).This test does not support</span><br><span class="hljs-comment">#&#x27;  pre-filtering of genes based on average difference (or percent detection rate)</span><br><span class="hljs-comment">#&#x27;  between cell groups. However, genes may be pre-filtered based on their</span><br><span class="hljs-comment">#&#x27;  minimum detection rate (min.pct) across both cell groups. To use this method,</span><br><span class="hljs-comment">#&#x27;  please install DESeq2, using the instructions at</span><br><span class="hljs-comment">#&#x27;  https://bioconductor.org/packages/release/bioc/html/DESeq2.html</span><br><span class="hljs-comment">#&#x27; &#125;</span><br></code></pre></td></tr></table></figure>



<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br><span class="line">94</span><br><span class="line">95</span><br><span class="line">96</span><br><span class="line">97</span><br><span class="line">98</span><br><span class="line">99</span><br><span class="line">100</span><br><span class="line">101</span><br><span class="line">102</span><br><span class="line">103</span><br><span class="line">104</span><br><span class="line">105</span><br><span class="line">106</span><br><span class="line">107</span><br><span class="line">108</span><br><span class="line">109</span><br><span class="line">110</span><br><span class="line">111</span><br><span class="line">112</span><br><span class="line">113</span><br><span class="line">114</span><br><span class="line">115</span><br><span class="line">116</span><br><span class="line">117</span><br><span class="line">118</span><br><span class="line">119</span><br><span class="line">120</span><br><span class="line">121</span><br><span class="line">122</span><br><span class="line">123</span><br><span class="line">124</span><br><span class="line">125</span><br></pre></td><td class="code"><pre><code class="hljs R">FindMarkers.default <span class="hljs-operator">&lt;-</span> <span class="hljs-keyword">function</span><span class="hljs-punctuation">(</span><br>  object<span class="hljs-punctuation">,</span><br>  slot <span class="hljs-operator">=</span> <span class="hljs-string">&quot;data&quot;</span><span class="hljs-punctuation">,</span><br>  cells.1 <span class="hljs-operator">=</span> <span class="hljs-literal">NULL</span><span class="hljs-punctuation">,</span><br>  cells.2 <span class="hljs-operator">=</span> <span class="hljs-literal">NULL</span><span class="hljs-punctuation">,</span><br>  features <span class="hljs-operator">=</span> <span class="hljs-literal">NULL</span><span class="hljs-punctuation">,</span><br>  logfc.threshold <span class="hljs-operator">=</span> <span class="hljs-number">0.1</span><span class="hljs-punctuation">,</span><br>  test.use <span class="hljs-operator">=</span> <span class="hljs-string">&quot;wilcox&quot;</span><span class="hljs-punctuation">,</span><br>  min.pct <span class="hljs-operator">=</span> <span class="hljs-number">0.01</span><span class="hljs-punctuation">,</span><br>  min.diff.pct <span class="hljs-operator">=</span> <span class="hljs-operator">-</span><span class="hljs-literal">Inf</span><span class="hljs-punctuation">,</span><br>  verbose <span class="hljs-operator">=</span> <span class="hljs-literal">TRUE</span><span class="hljs-punctuation">,</span><br>  only.pos <span class="hljs-operator">=</span> <span class="hljs-literal">FALSE</span><span class="hljs-punctuation">,</span><br>  max.cells.per.ident <span class="hljs-operator">=</span> <span class="hljs-literal">Inf</span><span class="hljs-punctuation">,</span><br>  random.seed <span class="hljs-operator">=</span> <span class="hljs-number">1</span><span class="hljs-punctuation">,</span><br>  latent.vars <span class="hljs-operator">=</span> <span class="hljs-literal">NULL</span><span class="hljs-punctuation">,</span><br>  min.cells.feature <span class="hljs-operator">=</span> <span class="hljs-number">3</span><span class="hljs-punctuation">,</span><br>  min.cells.group <span class="hljs-operator">=</span> <span class="hljs-number">3</span><span class="hljs-punctuation">,</span><br>  fc.results <span class="hljs-operator">=</span> <span class="hljs-literal">NULL</span><span class="hljs-punctuation">,</span><br>  densify <span class="hljs-operator">=</span> <span class="hljs-literal">FALSE</span><span class="hljs-punctuation">,</span><br>  ...<br><span class="hljs-punctuation">)</span> <span class="hljs-punctuation">&#123;</span><br>  ValidateCellGroups<span class="hljs-punctuation">(</span><br>    object <span class="hljs-operator">=</span> object<span class="hljs-punctuation">,</span><br>    cells.1 <span class="hljs-operator">=</span> cells.1<span class="hljs-punctuation">,</span><br>    cells.2 <span class="hljs-operator">=</span> cells.2<span class="hljs-punctuation">,</span><br>    min.cells.group <span class="hljs-operator">=</span> min.cells.group<br>  <span class="hljs-punctuation">)</span><br>  features <span class="hljs-operator">&lt;-</span> features <span class="hljs-operator">%||%</span> rownames<span class="hljs-punctuation">(</span>x <span class="hljs-operator">=</span> object<span class="hljs-punctuation">)</span><br>  <span class="hljs-comment"># reset parameters so no feature filtering is performed</span><br>  <span class="hljs-keyword">if</span> <span class="hljs-punctuation">(</span>test.use <span class="hljs-operator">%in%</span> DEmethods_noprefilter<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span> <span class="hljs-punctuation">&#123;</span><br>    features <span class="hljs-operator">&lt;-</span> rownames<span class="hljs-punctuation">(</span>x <span class="hljs-operator">=</span> object<span class="hljs-punctuation">)</span><br>    min.diff.pct <span class="hljs-operator">&lt;-</span> <span class="hljs-operator">-</span><span class="hljs-literal">Inf</span><br>    logfc.threshold <span class="hljs-operator">&lt;-</span> 0<br>  <span class="hljs-punctuation">&#125;</span><br>  <span class="hljs-comment"># feature selection (based on percentages)</span><br>  alpha.min <span class="hljs-operator">&lt;-</span> pmax<span class="hljs-punctuation">(</span>fc.results<span class="hljs-operator">$</span>pct.1<span class="hljs-punctuation">,</span> fc.results<span class="hljs-operator">$</span>pct.2<span class="hljs-punctuation">)</span><br>  <span class="hljs-built_in">names</span><span class="hljs-punctuation">(</span>x <span class="hljs-operator">=</span> alpha.min<span class="hljs-punctuation">)</span> <span class="hljs-operator">&lt;-</span> rownames<span class="hljs-punctuation">(</span>x <span class="hljs-operator">=</span> fc.results<span class="hljs-punctuation">)</span><br>  features <span class="hljs-operator">&lt;-</span> <span class="hljs-built_in">names</span><span class="hljs-punctuation">(</span>x <span class="hljs-operator">=</span> which<span class="hljs-punctuation">(</span>x <span class="hljs-operator">=</span> alpha.min <span class="hljs-operator">&gt;=</span> min.pct<span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span><br>  <span class="hljs-keyword">if</span> <span class="hljs-punctuation">(</span><span class="hljs-built_in">length</span><span class="hljs-punctuation">(</span>x <span class="hljs-operator">=</span> features<span class="hljs-punctuation">)</span> <span class="hljs-operator">==</span> <span class="hljs-number">0</span><span class="hljs-punctuation">)</span> <span class="hljs-punctuation">&#123;</span><br>    warning<span class="hljs-punctuation">(</span><span class="hljs-string">&quot;No features pass min.pct threshold; returning empty data.frame&quot;</span><span class="hljs-punctuation">)</span><br>    <span class="hljs-built_in">return</span><span class="hljs-punctuation">(</span>fc.results<span class="hljs-punctuation">[</span>features<span class="hljs-punctuation">,</span> <span class="hljs-punctuation">]</span><span class="hljs-punctuation">)</span><br>  <span class="hljs-punctuation">&#125;</span><br>  alpha.diff <span class="hljs-operator">&lt;-</span> alpha.min <span class="hljs-operator">-</span> pmin<span class="hljs-punctuation">(</span>fc.results<span class="hljs-operator">$</span>pct.1<span class="hljs-punctuation">,</span> fc.results<span class="hljs-operator">$</span>pct.2<span class="hljs-punctuation">)</span><br>  features <span class="hljs-operator">&lt;-</span> <span class="hljs-built_in">names</span><span class="hljs-punctuation">(</span><br>    x <span class="hljs-operator">=</span> which<span class="hljs-punctuation">(</span>x <span class="hljs-operator">=</span> alpha.min <span class="hljs-operator">&gt;=</span> min.pct <span class="hljs-operator">&amp;</span> alpha.diff <span class="hljs-operator">&gt;=</span> min.diff.pct<span class="hljs-punctuation">)</span><br>  <span class="hljs-punctuation">)</span><br>  <span class="hljs-keyword">if</span> <span class="hljs-punctuation">(</span><span class="hljs-built_in">length</span><span class="hljs-punctuation">(</span>x <span class="hljs-operator">=</span> features<span class="hljs-punctuation">)</span> <span class="hljs-operator">==</span> <span class="hljs-number">0</span><span class="hljs-punctuation">)</span> <span class="hljs-punctuation">&#123;</span><br>    warning<span class="hljs-punctuation">(</span><span class="hljs-string">&quot;No features pass min.diff.pct threshold; returning empty data.frame&quot;</span><span class="hljs-punctuation">)</span><br>    <span class="hljs-built_in">return</span><span class="hljs-punctuation">(</span>fc.results<span class="hljs-punctuation">[</span>features<span class="hljs-punctuation">,</span> <span class="hljs-punctuation">]</span><span class="hljs-punctuation">)</span><br>  <span class="hljs-punctuation">&#125;</span><br><br>  <span class="hljs-comment"># feature selection (based on logFC)</span><br>  <span class="hljs-keyword">if</span> <span class="hljs-punctuation">(</span>slot <span class="hljs-operator">!=</span> <span class="hljs-string">&quot;scale.data&quot;</span><span class="hljs-punctuation">)</span> <span class="hljs-punctuation">&#123;</span><br>    total.diff <span class="hljs-operator">&lt;-</span> fc.results<span class="hljs-punctuation">[</span><span class="hljs-punctuation">,</span> <span class="hljs-number">1</span><span class="hljs-punctuation">]</span> <span class="hljs-comment">#first column is logFC</span><br>    <span class="hljs-built_in">names</span><span class="hljs-punctuation">(</span>total.diff<span class="hljs-punctuation">)</span> <span class="hljs-operator">&lt;-</span> rownames<span class="hljs-punctuation">(</span>fc.results<span class="hljs-punctuation">)</span><br>    features.diff <span class="hljs-operator">&lt;-</span> <span class="hljs-keyword">if</span> <span class="hljs-punctuation">(</span>only.pos<span class="hljs-punctuation">)</span> <span class="hljs-punctuation">&#123;</span><br>      <span class="hljs-built_in">names</span><span class="hljs-punctuation">(</span>x <span class="hljs-operator">=</span> which<span class="hljs-punctuation">(</span>x <span class="hljs-operator">=</span> total.diff <span class="hljs-operator">&gt;=</span> logfc.threshold<span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span><br>    <span class="hljs-punctuation">&#125;</span> <span class="hljs-keyword">else</span> <span class="hljs-punctuation">&#123;</span><br>      <span class="hljs-built_in">names</span><span class="hljs-punctuation">(</span>x <span class="hljs-operator">=</span> which<span class="hljs-punctuation">(</span>x <span class="hljs-operator">=</span> <span class="hljs-built_in">abs</span><span class="hljs-punctuation">(</span>x <span class="hljs-operator">=</span> total.diff<span class="hljs-punctuation">)</span> <span class="hljs-operator">&gt;=</span> logfc.threshold<span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span><br>    <span class="hljs-punctuation">&#125;</span><br>    features <span class="hljs-operator">&lt;-</span> intersect<span class="hljs-punctuation">(</span>x <span class="hljs-operator">=</span> features<span class="hljs-punctuation">,</span> y <span class="hljs-operator">=</span> features.diff<span class="hljs-punctuation">)</span><br>    <span class="hljs-keyword">if</span> <span class="hljs-punctuation">(</span><span class="hljs-built_in">length</span><span class="hljs-punctuation">(</span>x <span class="hljs-operator">=</span> features<span class="hljs-punctuation">)</span> <span class="hljs-operator">==</span> <span class="hljs-number">0</span><span class="hljs-punctuation">)</span> <span class="hljs-punctuation">&#123;</span><br>      warning<span class="hljs-punctuation">(</span><span class="hljs-string">&quot;No features pass logfc.threshold threshold; returning empty data.frame&quot;</span><span class="hljs-punctuation">)</span><br>      <span class="hljs-built_in">return</span><span class="hljs-punctuation">(</span>fc.results<span class="hljs-punctuation">[</span>features<span class="hljs-punctuation">,</span> <span class="hljs-punctuation">]</span><span class="hljs-punctuation">)</span><br>    <span class="hljs-punctuation">&#125;</span><br>  <span class="hljs-punctuation">&#125;</span><br><br>  <span class="hljs-comment"># subsample cell groups if they are too large</span><br>  <span class="hljs-keyword">if</span> <span class="hljs-punctuation">(</span>max.cells.per.ident <span class="hljs-operator">&lt;</span> <span class="hljs-literal">Inf</span><span class="hljs-punctuation">)</span> <span class="hljs-punctuation">&#123;</span><br>    set.seed<span class="hljs-punctuation">(</span>seed <span class="hljs-operator">=</span> random.seed<span class="hljs-punctuation">)</span><br>    <span class="hljs-keyword">if</span> <span class="hljs-punctuation">(</span><span class="hljs-built_in">length</span><span class="hljs-punctuation">(</span>x <span class="hljs-operator">=</span> cells.1<span class="hljs-punctuation">)</span> <span class="hljs-operator">&gt;</span> max.cells.per.ident<span class="hljs-punctuation">)</span> <span class="hljs-punctuation">&#123;</span><br>      cells.1 <span class="hljs-operator">&lt;-</span> sample<span class="hljs-punctuation">(</span>x <span class="hljs-operator">=</span> cells.1<span class="hljs-punctuation">,</span> size <span class="hljs-operator">=</span> max.cells.per.ident<span class="hljs-punctuation">)</span><br>    <span class="hljs-punctuation">&#125;</span><br>    <span class="hljs-keyword">if</span> <span class="hljs-punctuation">(</span><span class="hljs-built_in">length</span><span class="hljs-punctuation">(</span>x <span class="hljs-operator">=</span> cells.2<span class="hljs-punctuation">)</span> <span class="hljs-operator">&gt;</span> max.cells.per.ident<span class="hljs-punctuation">)</span> <span class="hljs-punctuation">&#123;</span><br>      cells.2 <span class="hljs-operator">&lt;-</span> sample<span class="hljs-punctuation">(</span>x <span class="hljs-operator">=</span> cells.2<span class="hljs-punctuation">,</span> size <span class="hljs-operator">=</span> max.cells.per.ident<span class="hljs-punctuation">)</span><br>    <span class="hljs-punctuation">&#125;</span><br>    <span class="hljs-keyword">if</span> <span class="hljs-punctuation">(</span><span class="hljs-operator">!</span><span class="hljs-built_in">is.null</span><span class="hljs-punctuation">(</span>x <span class="hljs-operator">=</span> latent.vars<span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span> <span class="hljs-punctuation">&#123;</span><br>      latent.vars <span class="hljs-operator">&lt;-</span> latent.vars<span class="hljs-punctuation">[</span><span class="hljs-built_in">c</span><span class="hljs-punctuation">(</span>cells.1<span class="hljs-punctuation">,</span> cells.2<span class="hljs-punctuation">)</span><span class="hljs-punctuation">,</span> <span class="hljs-punctuation">,</span> drop <span class="hljs-operator">=</span> <span class="hljs-literal">FALSE</span><span class="hljs-punctuation">]</span><br>    <span class="hljs-punctuation">&#125;</span><br>  <span class="hljs-punctuation">&#125;</span><br>  <span class="hljs-keyword">if</span> <span class="hljs-punctuation">(</span>inherits<span class="hljs-punctuation">(</span>x <span class="hljs-operator">=</span> object<span class="hljs-punctuation">,</span> what <span class="hljs-operator">=</span> <span class="hljs-string">&quot;IterableMatrix&quot;</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">&#123;</span><br>    <span class="hljs-keyword">if</span><span class="hljs-punctuation">(</span>test.use <span class="hljs-operator">!=</span> <span class="hljs-string">&quot;wilcox&quot;</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">&#123;</span><br>      stop<span class="hljs-punctuation">(</span><span class="hljs-string">&quot;Differential expression with BPCells currently only supports the &#x27;wilcox&#x27; method.&quot;</span><span class="hljs-punctuation">,</span><br>           <span class="hljs-string">&quot; Please rerun with test.use = &#x27;wilcox&#x27;&quot;</span><span class="hljs-punctuation">)</span><br>    <span class="hljs-punctuation">&#125;</span><br>    data.use <span class="hljs-operator">&lt;-</span> object<span class="hljs-punctuation">[</span>features<span class="hljs-punctuation">,</span> <span class="hljs-built_in">c</span><span class="hljs-punctuation">(</span>cells.1<span class="hljs-punctuation">,</span> cells.2<span class="hljs-punctuation">)</span><span class="hljs-punctuation">,</span> drop <span class="hljs-operator">=</span> <span class="hljs-literal">FALSE</span><span class="hljs-punctuation">]</span><br>    groups <span class="hljs-operator">&lt;-</span> <span class="hljs-built_in">c</span><span class="hljs-punctuation">(</span><span class="hljs-built_in">rep</span><span class="hljs-punctuation">(</span><span class="hljs-string">&quot;foreground&quot;</span><span class="hljs-punctuation">,</span> <span class="hljs-built_in">length</span><span class="hljs-punctuation">(</span>cells.1<span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">,</span> <span class="hljs-built_in">rep</span><span class="hljs-punctuation">(</span><span class="hljs-string">&quot;background&quot;</span><span class="hljs-punctuation">,</span> <span class="hljs-built_in">length</span><span class="hljs-punctuation">(</span>cells.2<span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span><br>    de.results <span class="hljs-operator">&lt;-</span> suppressMessages<span class="hljs-punctuation">(</span><br>      BPCells<span class="hljs-operator">::</span>marker_features<span class="hljs-punctuation">(</span>data.use<span class="hljs-punctuation">,</span> group <span class="hljs-operator">=</span> groups<span class="hljs-punctuation">,</span> method <span class="hljs-operator">=</span> <span class="hljs-string">&quot;wilcoxon&quot;</span><span class="hljs-punctuation">)</span><br>    <span class="hljs-punctuation">)</span><br>    de.results <span class="hljs-operator">&lt;-</span> subset<span class="hljs-punctuation">(</span>de.results<span class="hljs-punctuation">,</span> de.results<span class="hljs-operator">$</span>foreground <span class="hljs-operator">==</span> <span class="hljs-string">&quot;foreground&quot;</span><span class="hljs-punctuation">)</span><br>    de.results <span class="hljs-operator">&lt;-</span> data.frame<span class="hljs-punctuation">(</span>feature <span class="hljs-operator">=</span> de.results<span class="hljs-operator">$</span>feature<span class="hljs-punctuation">,</span><br>                             p_val <span class="hljs-operator">=</span> de.results<span class="hljs-operator">$</span>p_val_raw<span class="hljs-punctuation">)</span><br>    rownames<span class="hljs-punctuation">(</span>de.results<span class="hljs-punctuation">)</span> <span class="hljs-operator">&lt;-</span> de.results<span class="hljs-operator">$</span>feature<br>    de.results<span class="hljs-operator">$</span>feature <span class="hljs-operator">&lt;-</span> <span class="hljs-literal">NULL</span><br>  <span class="hljs-punctuation">&#125;</span> <span class="hljs-keyword">else</span> <span class="hljs-punctuation">&#123;</span><br>    de.results <span class="hljs-operator">&lt;-</span> PerformDE<span class="hljs-punctuation">(</span><br>      object <span class="hljs-operator">=</span> object<span class="hljs-punctuation">,</span><br>      cells.1 <span class="hljs-operator">=</span> cells.1<span class="hljs-punctuation">,</span><br>      cells.2 <span class="hljs-operator">=</span> cells.2<span class="hljs-punctuation">,</span><br>      features <span class="hljs-operator">=</span> features<span class="hljs-punctuation">,</span><br>      test.use <span class="hljs-operator">=</span> test.use<span class="hljs-punctuation">,</span><br>      verbose <span class="hljs-operator">=</span> verbose<span class="hljs-punctuation">,</span><br>      min.cells.feature <span class="hljs-operator">=</span> min.cells.feature<span class="hljs-punctuation">,</span><br>      latent.vars <span class="hljs-operator">=</span> latent.vars<span class="hljs-punctuation">,</span><br>      densify <span class="hljs-operator">=</span> densify<span class="hljs-punctuation">,</span><br>      ...<br>    <span class="hljs-punctuation">)</span><br>  <span class="hljs-punctuation">&#125;</span><br>  de.results <span class="hljs-operator">&lt;-</span> cbind<span class="hljs-punctuation">(</span>de.results<span class="hljs-punctuation">,</span> fc.results<span class="hljs-punctuation">[</span>rownames<span class="hljs-punctuation">(</span>x <span class="hljs-operator">=</span> de.results<span class="hljs-punctuation">)</span><span class="hljs-punctuation">,</span> <span class="hljs-punctuation">,</span> drop <span class="hljs-operator">=</span> <span class="hljs-literal">FALSE</span><span class="hljs-punctuation">]</span><span class="hljs-punctuation">)</span><br>  <span class="hljs-keyword">if</span> <span class="hljs-punctuation">(</span>only.pos<span class="hljs-punctuation">)</span> <span class="hljs-punctuation">&#123;</span><br>    de.results <span class="hljs-operator">&lt;-</span> de.results<span class="hljs-punctuation">[</span>de.results<span class="hljs-punctuation">[</span><span class="hljs-punctuation">,</span> <span class="hljs-number">2</span><span class="hljs-punctuation">]</span> <span class="hljs-operator">&gt;</span> <span class="hljs-number">0</span><span class="hljs-punctuation">,</span> <span class="hljs-punctuation">,</span> drop <span class="hljs-operator">=</span> <span class="hljs-literal">FALSE</span><span class="hljs-punctuation">]</span><br>  <span class="hljs-punctuation">&#125;</span><br>  <span class="hljs-keyword">if</span> <span class="hljs-punctuation">(</span>test.use <span class="hljs-operator">%in%</span> DEmethods_nocorrect<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span> <span class="hljs-punctuation">&#123;</span><br>    de.results <span class="hljs-operator">&lt;-</span> de.results<span class="hljs-punctuation">[</span>order<span class="hljs-punctuation">(</span><span class="hljs-operator">-</span>de.results<span class="hljs-operator">$</span>power<span class="hljs-punctuation">,</span> <span class="hljs-operator">-</span>de.results<span class="hljs-punctuation">[</span><span class="hljs-punctuation">,</span> <span class="hljs-number">1</span><span class="hljs-punctuation">]</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">,</span> <span class="hljs-punctuation">]</span><br>  <span class="hljs-punctuation">&#125;</span> <span class="hljs-keyword">else</span> <span class="hljs-punctuation">&#123;</span><br>    de.results <span class="hljs-operator">&lt;-</span> de.results<span class="hljs-punctuation">[</span>order<span class="hljs-punctuation">(</span>de.results<span class="hljs-operator">$</span>p_val<span class="hljs-punctuation">,</span> <span class="hljs-operator">-</span><span class="hljs-built_in">abs</span><span class="hljs-punctuation">(</span>de.results<span class="hljs-operator">$</span>pct.1<span class="hljs-operator">-</span>de.results<span class="hljs-operator">$</span>pct.2<span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">,</span> <span class="hljs-punctuation">]</span><br>    de.results<span class="hljs-operator">$</span>p_val_adj <span class="hljs-operator">=</span> p.adjust<span class="hljs-punctuation">(</span><br>      p <span class="hljs-operator">=</span> de.results<span class="hljs-operator">$</span>p_val<span class="hljs-punctuation">,</span><br>      method <span class="hljs-operator">=</span> <span class="hljs-string">&quot;bonferroni&quot;</span><span class="hljs-punctuation">,</span><br>      n <span class="hljs-operator">=</span> nrow<span class="hljs-punctuation">(</span>x <span class="hljs-operator">=</span> object<span class="hljs-punctuation">)</span><br>    <span class="hljs-punctuation">)</span><br>  <span class="hljs-punctuation">&#125;</span><br>  <span class="hljs-built_in">return</span><span class="hljs-punctuation">(</span>de.results<span class="hljs-punctuation">)</span><br><span class="hljs-punctuation">&#125;</span><br></code></pre></td></tr></table></figure>

<p>FindMarkers使用了<code>presto</code>包，presto makes it fast and easy to run Wilcoxon rank sum test and auROC analysis on large datasets.</p>
<p>DE分析比较慢，如果想要提速度，需要调整的parameter有min.pct和logfc.threshold</p>
<blockquote>
<h2 id="Value"><a href="#Value" class="headerlink" title="Value"></a>Value</h2><p>data.frame with a ranked list of putative markers as rows, and associated statistics as columns (p-values, ROC score, etc., depending on the test used ()). The following columns are always present:<code>test.use</code></p>
<ul>
<li><code>avg_logFC</code>: log fold-chage of the average expression between the two groups. Positive values indicate that the gene is more highly expressed in the first group</li>
<li><code>pct.1</code>: The percentage of cells where the gene is detected in the first group</li>
<li><code>pct.2</code>: The percentage of cells where the gene is detected in the second group</li>
<li><code>p_val_adj</code>: Adjusted p-value, based on bonferroni correction using all genes in the dataset</li>
</ul>
</blockquote>
<p><img src="/../image/Seuratpictures%5Clog2FoldChange.png" srcset="/img/loading.gif" lazyload alt="log2FoldChange"></p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><code class="hljs R"><span class="hljs-comment"># find all markers distinguishing cluster 5 from clusters 0 and 3</span><br>cluster5.markers <span class="hljs-operator">&lt;-</span> FindMarkers<span class="hljs-punctuation">(</span>pbmc<span class="hljs-punctuation">,</span> ident.1 <span class="hljs-operator">=</span> <span class="hljs-number">5</span><span class="hljs-punctuation">,</span> ident.2 <span class="hljs-operator">=</span> <span class="hljs-built_in">c</span><span class="hljs-punctuation">(</span><span class="hljs-number">0</span><span class="hljs-punctuation">,</span> <span class="hljs-number">3</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span><br>head<span class="hljs-punctuation">(</span>cluster5.markers<span class="hljs-punctuation">,</span> n <span class="hljs-operator">=</span> <span class="hljs-number">5</span><span class="hljs-punctuation">)</span><br></code></pre></td></tr></table></figure>

<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><code class="hljs R"><span class="hljs-comment"># find markers for every cluster compared to all remaining cells, report only the positive</span><br><span class="hljs-comment"># ones</span><br>pbmc.markers <span class="hljs-operator">&lt;-</span> FindAllMarkers<span class="hljs-punctuation">(</span>pbmc<span class="hljs-punctuation">,</span> only.pos <span class="hljs-operator">=</span> <span class="hljs-literal">TRUE</span><span class="hljs-punctuation">)</span><br>pbmc.markers <span class="hljs-operator">%&gt;%</span><br>    group_by<span class="hljs-punctuation">(</span>cluster<span class="hljs-punctuation">)</span> <span class="hljs-operator">%&gt;%</span><br>    dplyr<span class="hljs-operator">::</span>filter<span class="hljs-punctuation">(</span>avg_log2FC <span class="hljs-operator">&gt;</span> <span class="hljs-number">1</span><span class="hljs-punctuation">)</span><br></code></pre></td></tr></table></figure>

<p>We include several tools for visualizing marker expression. <code>VlnPlot()</code> (shows expression probability distributions across clusters), and <code>FeaturePlot()</code> (visualizes feature expression on a tSNE or PCA plot) are our most commonly used visualizations. We also suggest exploring <code>RidgePlot()</code>, <code>CellScatter()</code>, and <code>DotPlot()</code> as additional methods to view your dataset.</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><code class="hljs R"><span class="hljs-comment"># you can plot raw counts as well</span><br>VlnPlot<span class="hljs-punctuation">(</span>pbmc<span class="hljs-punctuation">,</span> features <span class="hljs-operator">=</span> <span class="hljs-built_in">c</span><span class="hljs-punctuation">(</span><span class="hljs-string">&quot;NKG7&quot;</span><span class="hljs-punctuation">,</span> <span class="hljs-string">&quot;PF4&quot;</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">,</span> slot <span class="hljs-operator">=</span> <span class="hljs-string">&quot;counts&quot;</span><span class="hljs-punctuation">,</span> <span class="hljs-built_in">log</span> <span class="hljs-operator">=</span> <span class="hljs-literal">TRUE</span><span class="hljs-punctuation">)</span><br></code></pre></td></tr></table></figure>

<p><strong><code>slot = &quot;counts&quot;</code></strong>: 这个参数指定了从Seurat对象中提取原始计数数据。在这个例子中，我们想要绘制基因的原始计数数据，因此设置为”counts”。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><code class="hljs R">FeaturePlot<span class="hljs-punctuation">(</span>pbmc<span class="hljs-punctuation">,</span> features <span class="hljs-operator">=</span> <span class="hljs-built_in">c</span><span class="hljs-punctuation">(</span><span class="hljs-string">&quot;MS4A1&quot;</span><span class="hljs-punctuation">,</span> <span class="hljs-string">&quot;GNLY&quot;</span><span class="hljs-punctuation">,</span> <span class="hljs-string">&quot;CD3E&quot;</span><span class="hljs-punctuation">,</span> <span class="hljs-string">&quot;CD14&quot;</span><span class="hljs-punctuation">,</span> <span class="hljs-string">&quot;FCER1A&quot;</span><span class="hljs-punctuation">,</span> <span class="hljs-string">&quot;FCGR3A&quot;</span><span class="hljs-punctuation">,</span> <span class="hljs-string">&quot;LYZ&quot;</span><span class="hljs-punctuation">,</span> <span class="hljs-string">&quot;PPBP&quot;</span><span class="hljs-punctuation">,</span><br>    <span class="hljs-string">&quot;CD8A&quot;</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span><br></code></pre></td></tr></table></figure>

<p><img src="/../image/Seuratpictures%5Cunnamed-chunk-3-1.png" srcset="/img/loading.gif" lazyload alt="unnamed-chunk-3-1"></p>
<p><code>DoHeatmap()</code> generates an expression heatmap for given cells and features. In this case, we are plotting the top 20 markers (or all markers if less than 20) for each cluster.我怀疑这里写错了，应该是top 10 markers。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><code class="hljs R">pbmc.markers <span class="hljs-operator">%&gt;%</span><br>    group_by<span class="hljs-punctuation">(</span>cluster<span class="hljs-punctuation">)</span> <span class="hljs-operator">%&gt;%</span><br>    dplyr<span class="hljs-operator">::</span>filter<span class="hljs-punctuation">(</span>avg_log2FC <span class="hljs-operator">&gt;</span> <span class="hljs-number">1</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">%&gt;%</span><br>    slice_head<span class="hljs-punctuation">(</span>n <span class="hljs-operator">=</span> <span class="hljs-number">10</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">%&gt;%</span><br>    ungroup<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">-&gt;</span> top10<br>DoHeatmap<span class="hljs-punctuation">(</span>pbmc<span class="hljs-punctuation">,</span> features <span class="hljs-operator">=</span> top10<span class="hljs-operator">$</span>gene<span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span> NoLegend<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><br></code></pre></td></tr></table></figure>

<p><img src="/../image/Seuratpictures%5CclusterHeatmap-1.png" srcset="/img/loading.gif" lazyload alt="clusterHeatmap-1"></p>
<h2 id="六、细胞聚类注释"><a href="#六、细胞聚类注释" class="headerlink" title="六、细胞聚类注释"></a>六、细胞聚类注释</h2><p>seurat guide clusting tutorial写的比较简略。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><code class="hljs R">new.cluster.ids <span class="hljs-operator">&lt;-</span> <span class="hljs-built_in">c</span><span class="hljs-punctuation">(</span><span class="hljs-string">&quot;Naive CD4 T&quot;</span><span class="hljs-punctuation">,</span> <span class="hljs-string">&quot;CD14+ Mono&quot;</span><span class="hljs-punctuation">,</span> <span class="hljs-string">&quot;Memory CD4 T&quot;</span><span class="hljs-punctuation">,</span> <span class="hljs-string">&quot;B&quot;</span><span class="hljs-punctuation">,</span> <span class="hljs-string">&quot;CD8 T&quot;</span><span class="hljs-punctuation">,</span> <span class="hljs-string">&quot;FCGR3A+ Mono&quot;</span><span class="hljs-punctuation">,</span><br>    <span class="hljs-string">&quot;NK&quot;</span><span class="hljs-punctuation">,</span> <span class="hljs-string">&quot;DC&quot;</span><span class="hljs-punctuation">,</span> <span class="hljs-string">&quot;Platelet&quot;</span><span class="hljs-punctuation">)</span><br><span class="hljs-built_in">names</span><span class="hljs-punctuation">(</span>new.cluster.ids<span class="hljs-punctuation">)</span> <span class="hljs-operator">&lt;-</span> levels<span class="hljs-punctuation">(</span>pbmc<span class="hljs-punctuation">)</span><br>pbmc <span class="hljs-operator">&lt;-</span> RenameIdents<span class="hljs-punctuation">(</span>pbmc<span class="hljs-punctuation">,</span> new.cluster.ids<span class="hljs-punctuation">)</span><br>DimPlot<span class="hljs-punctuation">(</span>pbmc<span class="hljs-punctuation">,</span> reduction <span class="hljs-operator">=</span> <span class="hljs-string">&quot;umap&quot;</span><span class="hljs-punctuation">,</span> label <span class="hljs-operator">=</span> <span class="hljs-literal">TRUE</span><span class="hljs-punctuation">,</span> pt.size <span class="hljs-operator">=</span> <span class="hljs-number">0.5</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span> NoLegend<span class="hljs-punctuation">(</span><span class="hljs-punctuation">)</span><br></code></pre></td></tr></table></figure>

<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><code class="hljs R">library<span class="hljs-punctuation">(</span>ggplot2<span class="hljs-punctuation">)</span><br>plot <span class="hljs-operator">&lt;-</span> DimPlot<span class="hljs-punctuation">(</span>pbmc<span class="hljs-punctuation">,</span> reduction <span class="hljs-operator">=</span> <span class="hljs-string">&quot;umap&quot;</span><span class="hljs-punctuation">,</span> label <span class="hljs-operator">=</span> <span class="hljs-literal">TRUE</span><span class="hljs-punctuation">,</span> label.size <span class="hljs-operator">=</span> <span class="hljs-number">4.5</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span> xlab<span class="hljs-punctuation">(</span><span class="hljs-string">&quot;UMAP 1&quot;</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span> ylab<span class="hljs-punctuation">(</span><span class="hljs-string">&quot;UMAP 2&quot;</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span><br>    theme<span class="hljs-punctuation">(</span>axis.title <span class="hljs-operator">=</span> element_text<span class="hljs-punctuation">(</span>size <span class="hljs-operator">=</span> <span class="hljs-number">18</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">,</span> legend.text <span class="hljs-operator">=</span> element_text<span class="hljs-punctuation">(</span>size <span class="hljs-operator">=</span> <span class="hljs-number">18</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span> <span class="hljs-operator">+</span> guides<span class="hljs-punctuation">(</span>colour <span class="hljs-operator">=</span> guide_legend<span class="hljs-punctuation">(</span>override.aes <span class="hljs-operator">=</span> <span class="hljs-built_in">list</span><span class="hljs-punctuation">(</span>size <span class="hljs-operator">=</span> <span class="hljs-number">10</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span><span class="hljs-punctuation">)</span><br>ggsave<span class="hljs-punctuation">(</span>filename <span class="hljs-operator">=</span> <span class="hljs-string">&quot;../output/images/pbmc3k_umap.jpg&quot;</span><span class="hljs-punctuation">,</span> height <span class="hljs-operator">=</span> <span class="hljs-number">7</span><span class="hljs-punctuation">,</span> width <span class="hljs-operator">=</span> <span class="hljs-number">12</span><span class="hljs-punctuation">,</span> plot <span class="hljs-operator">=</span> plot<span class="hljs-punctuation">,</span> quality <span class="hljs-operator">=</span> <span class="hljs-number">50</span><span class="hljs-punctuation">)</span><br></code></pre></td></tr></table></figure>

<p><img src="/../image/Seuratpictures%5Clabelplot-1.png" srcset="/img/loading.gif" lazyload alt="labelplot-1"></p>

                
              </div>
            
            <hr/>
            <div>
              <div class="post-metas my-3">
  
  
    <div class="post-meta">
      <i class="iconfont icon-tags"></i>
      
        <a href="/tags/R/" class="print-no-link">#R</a>
      
        <a href="/tags/Seurat/" class="print-no-link">#Seurat</a>
      
    </div>
  
</div>


              
  

  <div class="license-box my-3">
    <div class="license-title">
      <div>Learning Seurat - Guided Clustering Tutorial</div>
      <div>http://example.com/2024/09/03/20240903Seurat - Guided Clustering Tutorial/</div>
    </div>
    <div class="license-meta">
      
        <div class="license-meta-item">
          <div>作者</div>
          <div>Jingxi Qin</div>
        </div>
      
      
        <div class="license-meta-item license-meta-date">
          <div>发布于</div>
          <div>2024年9月3日</div>
        </div>
      
      
      
        <div class="license-meta-item">
          <div>许可协议</div>
          <div>
            
              
              
                <a class="print-no-link" target="_blank" href="https://creativecommons.org/licenses/by/4.0/">
                  <span class="hint--top hint--rounded" aria-label="BY - 署名">
                    <i class="iconfont icon-cc-by"></i>
                  </span>
                </a>
              
            
          </div>
        </div>
      
    </div>
    <div class="license-icon iconfont"></div>
  </div>



              
                <div class="post-prevnext my-3">
                  <article class="post-prev col-6">
                    
                    
                      <a href="/2024/09/03/20240903R%E7%BC%96%E7%A8%8B/" title="R编程">
                        <i class="iconfont icon-arrowleft"></i>
                        <span class="hidden-mobile">R编程</span>
                        <span class="visible-mobile">上一篇</span>
                      </a>
                    
                  </article>
                  <article class="post-next col-6">
                    
                    
                      <a href="/2024/09/03/20240903Seurat%E9%A2%84%E5%A4%84%E7%90%86/" title="Learning Seurat - 预处理">
                        <span class="hidden-mobile">Learning Seurat - 预处理</span>
                        <span class="visible-mobile">下一篇</span>
                        <i class="iconfont icon-arrowright"></i>
                      </a>
                    
                  </article>
                </div>
              
            </div>

            
          </article>
        </div>
      </div>
    </div>

    <div class="side-col d-none d-lg-block col-lg-2">
      
  <aside class="sidebar" style="margin-left: -1rem">
    <div id="toc">
  <p class="toc-header">
    <i class="iconfont icon-list"></i>
    <span>目录</span>
  </p>
  <div class="toc-body" id="toc-body"></div>
</div>



  </aside>


    </div>
  </div>
</div>





  



  



  



  



  







    

    
      <a id="scroll-top-button" aria-label="TOP" href="#" role="button">
        <i class="iconfont icon-arrowup" aria-hidden="true"></i>
      </a>
    

    
      <div class="modal fade" id="modalSearch" tabindex="-1" role="dialog" aria-labelledby="ModalLabel"
     aria-hidden="true">
  <div class="modal-dialog modal-dialog-scrollable modal-lg" role="document">
    <div class="modal-content">
      <div class="modal-header text-center">
        <h4 class="modal-title w-100 font-weight-bold">搜索</h4>
        <button type="button" id="local-search-close" class="close" data-dismiss="modal" aria-label="Close">
          <span aria-hidden="true">&times;</span>
        </button>
      </div>
      <div class="modal-body mx-3">
        <div class="md-form mb-5">
          <input type="text" id="local-search-input" class="form-control validate">
          <label data-error="x" data-success="v" for="local-search-input">关键词</label>
        </div>
        <div class="list-group" id="local-search-result"></div>
      </div>
    </div>
  </div>
</div>

    

    
  </main>

  <footer>
    <div class="footer-inner">
  
    <div class="footer-content">
       <a href="https://hexo.io" target="_blank" rel="nofollow noopener"><span>Hexo</span></a> <i class="iconfont icon-love"></i> <a href="https://github.com/fluid-dev/hexo-theme-fluid" target="_blank" rel="nofollow noopener"><span>Fluid</span></a> 
    </div>
  
  
    <div class="statistics">
  
  

  
    
      <span id="busuanzi_container_site_pv" style="display: none">
        总访问量 
        <span id="busuanzi_value_site_pv"></span>
         次
      </span>
    
    
      <span id="busuanzi_container_site_uv" style="display: none">
        总访客数 
        <span id="busuanzi_value_site_uv"></span>
         人
      </span>
    
    

  

</div>

  
  
</div>

  </footer>

  <!-- Scripts -->
  
  <script  src="https://lib.baomitu.com/nprogress/0.2.0/nprogress.min.js" ></script>
  <link  rel="stylesheet" href="https://lib.baomitu.com/nprogress/0.2.0/nprogress.min.css" />

  <script>
    NProgress.configure({"showSpinner":false,"trickleSpeed":100})
    NProgress.start()
    window.addEventListener('load', function() {
      NProgress.done();
    })
  </script>


<script  src="https://lib.baomitu.com/jquery/3.6.4/jquery.min.js" ></script>
<script  src="https://lib.baomitu.com/twitter-bootstrap/4.6.1/js/bootstrap.min.js" ></script>
<script  src="/js/events.js" ></script>
<script  src="/js/plugins.js" ></script>


  <script  src="https://lib.baomitu.com/typed.js/2.0.12/typed.min.js" ></script>
  <script>
    (function (window, document) {
      var typing = Fluid.plugins.typing;
      var subtitle = document.getElementById('subtitle');
      if (!subtitle || !typing) {
        return;
      }
      var text = subtitle.getAttribute('data-typed-text');
      
        typing(text);
      
    })(window, document);
  </script>




  
    <script  src="/js/img-lazyload.js" ></script>
  




  
<script>
  Fluid.utils.createScript('https://lib.baomitu.com/tocbot/4.20.1/tocbot.min.js', function() {
    var toc = jQuery('#toc');
    if (toc.length === 0 || !window.tocbot) { return; }
    var boardCtn = jQuery('#board-ctn');
    var boardTop = boardCtn.offset().top;

    window.tocbot.init(Object.assign({
      tocSelector     : '#toc-body',
      contentSelector : '.markdown-body',
      linkClass       : 'tocbot-link',
      activeLinkClass : 'tocbot-active-link',
      listClass       : 'tocbot-list',
      isCollapsedClass: 'tocbot-is-collapsed',
      collapsibleClass: 'tocbot-is-collapsible',
      scrollSmooth    : true,
      includeTitleTags: true,
      headingsOffset  : -boardTop,
    }, CONFIG.toc));
    if (toc.find('.toc-list-item').length > 0) {
      toc.css('visibility', 'visible');
    }

    Fluid.events.registerRefreshCallback(function() {
      if ('tocbot' in window) {
        tocbot.refresh();
        var toc = jQuery('#toc');
        if (toc.length === 0 || !tocbot) {
          return;
        }
        if (toc.find('.toc-list-item').length > 0) {
          toc.css('visibility', 'visible');
        }
      }
    });
  });
</script>


  <script src=https://lib.baomitu.com/clipboard.js/2.0.11/clipboard.min.js></script>

  <script>Fluid.plugins.codeWidget();</script>


  
<script>
  Fluid.utils.createScript('https://lib.baomitu.com/anchor-js/5.0.0/anchor.min.js', function() {
    window.anchors.options = {
      placement: CONFIG.anchorjs.placement,
      visible  : CONFIG.anchorjs.visible
    };
    if (CONFIG.anchorjs.icon) {
      window.anchors.options.icon = CONFIG.anchorjs.icon;
    }
    var el = (CONFIG.anchorjs.element || 'h1,h2,h3,h4,h5,h6').split(',');
    var res = [];
    for (var item of el) {
      res.push('.markdown-body > ' + item.trim());
    }
    if (CONFIG.anchorjs.placement === 'left') {
      window.anchors.options.class = 'anchorjs-link-left';
    }
    window.anchors.add(res.join(', '));

    Fluid.events.registerRefreshCallback(function() {
      if ('anchors' in window) {
        anchors.removeAll();
        var el = (CONFIG.anchorjs.element || 'h1,h2,h3,h4,h5,h6').split(',');
        var res = [];
        for (var item of el) {
          res.push('.markdown-body > ' + item.trim());
        }
        if (CONFIG.anchorjs.placement === 'left') {
          anchors.options.class = 'anchorjs-link-left';
        }
        anchors.add(res.join(', '));
      }
    });
  });
</script>


  
<script>
  Fluid.utils.createScript('https://lib.baomitu.com/fancybox/3.5.7/jquery.fancybox.min.js', function() {
    Fluid.plugins.fancyBox();
  });
</script>


  <script>Fluid.plugins.imageCaption();</script>

  <script  src="/js/local-search.js" ></script>

  <script defer src="https://busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js" ></script>





<!-- 主题的启动项，将它保持在最底部 -->
<!-- the boot of the theme, keep it at the bottom -->
<script  src="/js/boot.js" ></script>


  

  <noscript>
    <div class="noscript-warning">博客在允许 JavaScript 运行的环境下浏览效果更佳</div>
  </noscript>
</body>
</html>
